# AD transcriptomic profiling in time and space

## Load packages and environment variable

```{r load packages, echo=FALSE, message=FALSE, warning=FALSE}
pacman::p_load(tibble, dplyr, tidyr, readr, stringr, ggplot2, purrr, 
               tidybulk, tidySummarizedExperiment, 
               ComplexHeatmap, ggpubr,
               BiocParallel, Biobase, variancePartition,
               DESeq2, limma, edgeR, MuSiC, convert, 
               patchwork, lme4, lmerTest, broom.mixed,
               vegan, tabula)

knitr::opts_chunk$set(message = FALSE, 
                      warning = FALSE, 
                      echo = FALSE, 
                      fig.align = 'center')
```

```{r environment variable}
pca_color <- c("LS" = "#eb2d0c", "NL" = "#eb8b9b", "HC" = "#91cf60")
core_n <- future::availableCores()
register(MulticoreParam(ifelse(core_n <= 8, core_n - 2, core_n / 2)))
```

```{r load data}
# read raw counts ---------------------------------------
counts <-
  data.table::fread("../multiomics-ad-transcriptomics/data/counts/counts_gene_name_redo.txt.gz") %>% 
  mutate(Chr = Chr %>% str_extract("chr\\w{1,}(?=;)|chr\\w{1,}$"), #keeps only 1
         Strand = Strand %>% str_extract("\\+|\\-"),               #keeps only 1
         Start = Start %>% str_extract("\\d{1,}(?=;)|\\d{1,}$"),   #keeps only 1
         End = End %>% str_extract("\\d{1,}(?=;)|\\d{1,}$")) %>%   #keeps only 1
  filter(!Geneid %>% 
           stringr::str_detect("\\d{1,}P$|\\d{1,}P\\d{1,}$|\\.|-AS\\d{1}|-DT")) %>% # apply gene filtering (pseudo genes, antisense genese, DT genes)
  dplyr::rename_with(str_extract, starts_with("/home/projects/"), pattern = "NG-[:graph:]{1,}$")

# lib to merge -------------
lib_id <- colnames(counts)[-1:-6] %>% str_extract("lib\\d{1,}") # extract lib id
lib_id_t <- table(lib_id)
lib_id_merge <- lib_id_t[(lib_id_t != 1)] %>% rownames() #lib_ids to be merged

# merge technical replicates -------------------------------------------
source("../multiomics-ad-transcriptomics/R/helper.R")

counttable_LibMerged <- 
  bind_cols(counts[, 1:6], 
            counttable_merge_library_fun(counts %>% dplyr::select(-1:-6), lib_id_merge)) 

# clinical records ------------------------
clinical_records <- 
  readr::read_csv("../multiomics-ad-transcriptomics/data/LEO-AD-metadata.csv") %>% 
  mutate(id = id %>% str_replace("-", "_"),
         JOIN_ID = paste(id, paste0("0",visit_no), sep="_"))

extensive_meta <-
  readr::read_csv("../multiomics-ad-transcriptomics/data/RNAseq_sample_annotation(extensive).csv") %>% 
  mutate(
    skin_type_fitzpatrick_scale = skin_type_fitzpatrick_scale %>% as.factor()
      )
```

```{r SummarizedExperiment}
# revised biopsy area by Tanja
revised_biopsy_area <- 
  readr::read_csv("../multiomics-ad-transcriptomics/data/tanja_replicate_location.csv") %>% 
  mutate(biopsy_area = biopsy_area %>% str_replace_all(" ", "_"))

# revise space infomation
biopsy_area <- 
  extensive_meta %>% 
  select(BAM_ID, biopsy_area_old = biopsy_area) %>% 
  left_join(revised_biopsy_area %>% select(BAM_ID, biopsy_area)) %>% 
  mutate(biopsy_area = ifelse(is.na(biopsy_area), biopsy_area_old, biopsy_area)) %>% 
  mutate(biopsy_area = ifelse(biopsy_area == "lÃ¦g", "leg", biopsy_area)) %>% 
  select(-biopsy_area_old) %>% 
  add_case(BAM_ID = "NG-23830_01_AD_07_BI_LS_01_E9_lib390881_6751_1.bam",
           biopsy_area = "arm")

# se col_data
col_data <-
  tibble(
    # definingn metadata --------------------------------------
    BAM_ID = colnames(counttable_LibMerged)[-1:-6]) %>% 
    mutate(sequencing_batch = BAM_ID %>% str_extract("NG-\\d{1,}"),
           group = BAM_ID %>% str_extract("AD|CO") %>% 
               forcats::fct_relevel(c("CO", "AD")),
           subject = BAM_ID %>% str_extract("(AD|CO)_\\d{2,}"),
           visit = BAM_ID %>% str_extract("\\d{2,}_(AD|CO)") %>% str_extract("\\d{2,}"),
           skin_type = BAM_ID %>% str_replace("TS", "LS") %>% str_extract("NN|NL|LS") %>% forcats::fct_relevel(c("NN", "NL", "LS")),
           replicate_ID = BAM_ID %>% str_replace("TS_BI", "BI_LS") %>% str_extract("(NN|NL|LS)_\\d{2,}") %>% str_extract("\\d{2,}"),
           library_ID = BAM_ID %>% str_extract("lib\\d{1,}"),
           sequencing_id = BAM_ID %>% str_extract("\\d{4}_(1|2)"),
           JOIN_ID = base::paste(subject, visit, sep = "_")) %>%
    left_join(clinical_records %>% select(id, gender) %>% filter(!is.na(gender)), by=c("subject"="id")) %>%
    left_join(clinical_records %>% select(JOIN_ID, scorad, scorad_objective, easi_total_score, date_visit)) %>% 
    select(-JOIN_ID) %>% 
    mutate(gender = gender %>% forcats::fct_relevel(c("female", "male"))) %>% 
    left_join(extensive_meta %>% select(BAM_ID, rna_quality, visit_quarter)) %>% 
    left_join(biopsy_area)

# Summarized Experiment--------------------
se <- SummarizedExperiment::SummarizedExperiment(
  assays = list(counts = counttable_LibMerged %>% select(-1:-6) %>% as.matrix),
  colData = col_data
)
names(se) <- counttable_LibMerged$Geneid
colData(se)$biopsy_area <- colData(se)$biopsy_area %>% 
  str_extract("arm|back_of_knee|elbow|feet|hand|leg|wrist")
colData(se) <- 
  DataFrame(
  colData(se) %>% 
    as.data.frame() %>% 
    mutate(skin_type = as.character(skin_type),
           skin_type = ifelse(skin_type == "NN", "HC", skin_type),
           skin_type = forcats::fct_relevel(skin_type, c("HC", "NL", "LS")))
  )
```

```{r filtering}
# filter out samples ----------------
se <- se[, se$BAM_ID != (assay(se) %>% colSums() %>% which.min %>% names)] # 392
se <- se[, !se$library_ID %in% c("lib390174", "lib390716")] # 390
```

```{r revise some sp info}
index_ad03v01ls01 <-
  which(colData(se)$library_ID == "lib390797")
colData(se)$skin_type[index_ad03v01ls01] <- "NL"

index_ad03v01nl01 <-
  which(colData(se)$library_ID == "lib390169")
colData(se)$skin_type[index_ad03v01nl01] <- "LS"

index_ad13v03nl01 <-
  which(colData(se)$library_ID == "lib390877")
colData(se)$visit[index_ad13v03nl01] <- "05"
colData(se)$skin_type[index_ad13v03nl01] <- "LS"
colData(se)$replicate_ID[index_ad13v03nl01] <- "02"

index_ad13v05ls02 <-
  which(colData(se)$library_ID == "lib390170")
colData(se)$visit[index_ad13v05ls02] <- "03"
colData(se)$skin_type[index_ad13v05ls02] <- "NL"
colData(se)$replicate_ID[index_ad13v05ls02] <- "01"

# se %>% write_rds("data/se_raw.rds")
```

```{r Sample QC}
sample_qc_record_edited <-
  readr::read_csv("data/sample_qc_record.csv")
```

```{r}
sample_qc_record_edited %>% filter(exclude == TRUE) %>% 
  group_by(skin_type) %>% summarise(n = n())
```

```{r}
sample_qc_record_edited %>% filter(exclude == TRUE) %>% 
  group_by(subject, visit, skin_type) %>% 
  summarise(n=n()) %>% arrange(-n)
```
```{r}
sample_qc_record_rna_conc <- 
  sample_qc_record_edited %>%
  left_join(
list.files("../multiomics-ad-transcriptomics/data/qc_record/", full.names = T) %>% 
    lapply(readxl::read_excel) %>% purrr::reduce(bind_rows) %>% 
    mutate(subject = `Customer Sample Name` %>% str_extract("(AD|CO)_\\d{2}"),
           visit = `Customer Sample Name` %>% str_extract("^\\d{2}"),
           skin_type = `Customer Sample Name` %>% str_extract("(LS|NL|NN)"),
           replicate_ID = `Customer Sample Name` %>% str_extract("(?<=(LS|NL|NN)_)(01|02)(?=_)"))
                     ) %>% 
      select(subject, visit, skin_type, replicate_ID, `FA Concentration [ng/ul]`)

tibble(
  sample_qc_record_rna_conc %>% group_by(subject) %>% summarise(n_rnac=n())
) %>% 
  left_join(sample_qc_record_edited %>% group_by(subject) %>% summarise(n_qc=n())) %>% 
  mutate(equal = n_rnac == n_qc)


```



```{r exclude}
exclude_BAM <- 
  colData(se) %>% as_tibble() %>% 
  select(BAM_ID, subject, visit, skin_type, replicate_ID) %>%
  left_join(sample_qc_record_edited) %>% 
  filter(exclude == TRUE) %>% pull(BAM_ID)
```

```{r generate new se}
se <- se[, !se$BAM_ID %in% exclude_BAM]

g_non_pseudo <- 
  rownames(se) %>% 
  gprofiler2::gconvert() %>% 
  filter(!description %>% str_detect("pseudogene")) %>% 
  pull(input) %>% unique()

se <-
  se[g_non_pseudo,]

# make a cope of a non-filtered se
se_prifilt <- se

# filter out genes 
se <- 
  se_prifilt %>% 
  keep_abundant(minimum_proportion = 0) %>%
  scale_abundance()
```

```{r generate KRTAP}
se_KRTAP <- se_prifilt %>% 
  keep_abundant(factor_of_interest = skin_type, minimum_proportion = .1) %>%
  scale_abundance()
```

```{r KRTAP variance}
KRTAP_hair <-
  se_KRTAP %>% 
  filter(feature %>% str_detect("KRTAP")) %>% 
  group_by(feature) %>% 
  summarise(mean_expr = mean(counts), 
            sd_expr = sd(counts)) %>%   
  mutate(expr = ifelse(mean_expr >= 10, TRUE, FALSE))

se_KRTAP_filt <- 
  se_KRTAP %>% 
  filter(feature %in% (KRTAP_hair %>% filter(expr == TRUE) %>% pull(feature)))
```

```{r write out data, eval=FALSE}
readr::write_rds(se, "data/se.rds")
readr::write_rds(se_KRTAP_filt, "data/se_krtap_all24.rds")
```

## Upload data to GEO

```{r checksum files, eval=FALSE}
md5sum <- 
  readr::read_delim("data/geo/geo_md5sum.txt", delim = "  ", 
                  col_names = c("md5sum", "file_name")) %>% 
  mutate(file_name=file_name %>% str_remove("read/"),
         lib_id = file_name %>% str_extract("lib\\d{1,}"))
```

```{r metadata for GEO, eval=FALSE}
geo_metadata <- 
  colData(se) %>% as_tibble() %>% 
  mutate(title = paste(visit, subject, skin_type, replicate_ID,
                       sep = "_"),
         `source name` = "skin biopsy",
         organism = "Homo sapians") %>% 
  select(BAM_ID,
         title, 
         `source name`,
         organism,
         subject,
         skin_type,
         gender,
         scorad,
         easi_total_score,
         visit, date_visit,
         biopsy_area,
         library_ID)

geo_metadata_n <-
  geo_metadata %>% 
  # select(title, library_ID) %>%
  left_join(md5sum, by = c("library_ID" = "lib_id")) %>% 
  group_by(title) %>% nest() %>% 
  mutate(raw_file = data %>% map("file_name"),
         md5sum = data %>% map("md5sum"),
         data = data %>% map(~ select(.x, -md5sum, -file_name))) %>% 
  unnest(cols = data) %>% 
  distinct()

geo_metadata_n %>% 
  openxlsx::write.xlsx("data/geo/geo_meta.xlsx")

geo_md5sum <- 
  geo_metadata_n %>% 
  ungroup %>% 
  unnest(cols = c(raw_file, md5sum)) %>% 
  select(raw_file, md5sum)
openxlsx::write.xlsx(geo_md5sum, "data/geo/geo_md5sum.xlsx")

count_matrix <-
  assay(se, 1) %>% as_tibble(rownames = "gene_name")
colnames(count_matrix)[-1] <- geo_metadata$title
readr::write_csv(count_matrix, "data/geo/count_matrix.csv")

count_matrix_normalized <-
  assay(se, 2) %>% as_tibble(rownames = "gene_name")
colnames(count_matrix_normalized)[-1] <- geo_metadata$title
readr::write_csv(count_matrix_normalized, "data/geo/count_matrix_normalized.csv")

```

```{r paired, eval=FALSE}
geo_md5sum %>% 
  select(raw_file) %>% 
  mutate(raw_file = raw_file %>% str_remove("_(1|2).fastq.gz")) %>% 
  distinct() %>% 
  mutate(pair_1 = paste0(raw_file, "_1.fastq.gz"),
         pair_2 = paste0(raw_file, "_2.fastq.gz")) %>% 
  select(contains("pair")) %>% 
  openxlsx::write.xlsx("data/geo/geo_paired.xlsx")
```


## Baseline characteristics (Table S1 and Table 1)

```{r baseline-statistics-table}
extensive_meta <- 
  readr::read_csv("../multiomics-ad-transcriptomics/data/RNAseq_sample_annotation(extensive).csv") 

baseline_data <-  
  extensive_meta %>% filter(visit == "01") %>% 
  select(subject, 
         group, 
         # t test
         age, 
         bmi, 
         contains("blood"),
         # fisher's exact test
         gender, 
         smoker,
         smoker_ever,
         animal,
         family_eczema,
         family_atopy,
         other_atopy,
         other_diagnoses,
         prick_test_result,
         # Mann-Whitney-Wilcoxon Test
         education,
         drinks_per_week,
         skin_type_fitzpatrick_scale,
         grow_up_where,
         # NOT FOR COMPARISON
         easi_total_score, scorad, 
         scorad_objective) %>% 
  distinct() %>% 
  mutate(ige_150h = ifelse(blood_ige > 150, TRUE, FALSE),
         gender_female = ifelse(gender == "female", T, F),
         family_eczema = ifelse(family_eczema == "-", 
                                F,
                                family_eczema),
         family_eczema = as.logical(family_eczema),
         other_atopy = ifelse(other_atopy == "-",
                              F,
                              other_atopy),
         other_atopy = as.logical(other_atopy),
         prick_test_result = case_when(prick_test_result == "POS" ~ TRUE,
                                       prick_test_result == "NEG" ~ FALSE,
                                       TRUE ~ NA),
         education = ifelse(education == "meidium_higher_education",
                            "medium_higher_education",
                            education),
         education = case_when(education == "primary_school" ~ 0,
                               education == "high_school" ~ 1,
                               education == "short_higher_education" ~ 2,
                               education == "medium_higher_education" ~ 3,
                               education == "long_higher_education" ~ 4),
         drinks_per_week = case_when(drinks_per_week == "0" ~ 0,
                                     drinks_per_week == "1_7" ~ 1,
                                     drinks_per_week == "7_14" ~ 2,
                                     drinks_per_week == "15_" ~ 3))

# Mann-Whitney-Wilcoxon Test
wilcox_test_d <-
  baseline_data %>% 
  select(subject, group, 
         education, skin_type_fitzpatrick_scale, drinks_per_week) %>% 
  pivot_longer(cols = c("education", "skin_type_fitzpatrick_scale", "drinks_per_week"),
               names_to = "parameter") %>% 
  group_by(parameter) %>% nest() %>% 
  mutate(p_value = map_dbl(data, function(data){
    CO <- data %>% filter(group == "CO")
    AD <- data %>% filter(group == "AD")
    wilcox.test(CO$value, AD$value, exact = F)$p.value
  }))

# t test
t_test_d <- 
  baseline_data %>% 
  select(subject, group, contains("blood"), age, bmi) %>% 
  pivot_longer(cols = !c("subject", "group"), 
               names_to = "parameter") %>% 
  group_by(parameter) %>% nest() %>% 
  mutate(p_value = map(data, function(data){
    CO <- data %>% filter(group == "CO")
    AD <- data %>% filter(group == "AD")
    t.test(CO$value, AD$value, na.rm = T)$p.value
  }))


# fisher's exact test
fisher_test_d <-
  baseline_data %>% 
  select(subject, group, gender_female, smoker, smoker_ever,
         animal, family_eczema, family_atopy, 
         other_atopy, other_diagnoses,
         prick_test_result) %>% 
  pivot_longer(cols = !c("subject", "group"),
               names_to = "parameter") %>% 
  group_by(parameter) %>% nest() %>% 
  mutate(
    AD = map_dbl(data, function(data){
      data %>% filter(group == "AD", value == TRUE) %>% nrow()
    }),
    CO = map_dbl(data, function(data){
      data %>% filter(group == "CO", value == TRUE) %>% nrow()
    }),
    p_value = map_dbl(data, function(data){
      test <- data %>% select(-subject) %>% table %>% fisher.test()
      test$p.value
      })
    )


clinic_summary <- 
 baseline_data %>% select(-subject) %>% group_by(group) %>% 
  summarize(no_subject = n(),
            female_percent = mean(gender == "female") * 100,
            age_mean = mean(age), age_sd = sd(age), 
            bmi_mean = mean(bmi), bmi_sd = sd(bmi), 
            blood_ige_mean = mean(blood_ige, na.rm = TRUE), blood_ige_sd = sd(blood_ige, na.rm = TRUE),
            blood_leukocyte_mean = mean(blood_leukocyte, na.rm=T), 
            blood_leukocyte_sd = sd(blood_leukocyte, na.rm=T),
            blood_lymphocyte_mean = mean(blood_lymphocyte, na.rm=T), 
            blood_lymphocyte_sd = sd(blood_lymphocyte, na.rm=T),
            blood_monocyte_mean = mean(blood_monocyte, na.rm=T),
            blood_monocyte_sd = sd(blood_monocyte, na.rm=T),
            blood_neutrophil_mean = mean(blood_neutrophil, na.rm = T),
            blood_neutrophil_sd = sd(blood_neutrophil, na.rm = T),
            blood_eosinophil_mean = mean(blood_eosinophil, na.rm = T),
            blood_eosinophil_sd = sd(blood_eosinophil, na.rm = T),
            blood_basophil_mean = mean(blood_basophil, na.rm = T),
            blood_basophil_sd = sd(blood_basophil, na.rm = T),
            easi_mean = mean(easi_total_score), easi_sd = sd(easi_total_score), 
            scorad_mean = mean(scorad, na.rm = TRUE), scorad_sd = sd(scorad, na.rm = TRUE),
            oscorad_mean = mean(scorad_objective, na.rm = TRUE), 
            oscorad_sd = sd(scorad_objective, na.rm = TRUE),
            ige_mean = mean(blood_ige, na.rm = TRUE), 
            ige_sd = sd(blood_ige, na.rm = TRUE),
            ige_median = median(blood_ige, na.rm = TRUE),
            ige_150h = sum(blood_ige > 150, na.rm = TRUE))

p_value <- 
  tibble(
    `Female%` = baseline_data %>% select(gender, group) %>% 
      table %>% fisher.test() %>% broom::tidy() %>% pull(p.value),
    `Blood_Ige > 150` = baseline_data %>% select(group, ige_150h) %>% 
      table %>% fisher.test() %>% broom::tidy() %>% pull(p.value),
    age = t.test(x = baseline_data %>% filter(group == "AD") %>% select(age),
                 y = baseline_data %>% filter(group == "CO") %>% select(age))$p.value,
    bmi = t.test(x = baseline_data %>% filter(group == "AD") %>% select(bmi),
                 y = baseline_data %>% filter(group == "CO") %>% select(bmi))$p.value,
    `blood_ige` = t.test(x = baseline_data %>% filter(group == "AD") %>% select(blood_ige),
                         y = baseline_data %>% filter(group == "CO") %>% select(blood_ige))$p.value) %>% 
  mutate_all(round, digits = 2) %>% mutate_all(as.character)

clinic_summary_t <-
  clinic_summary %>% 
  mutate(group = paste0(group, " (n=", no_subject,")"),
         `Female%` = female_percent %>% round(),
         age = paste(age_mean %>% round(1) , "Â±", age_sd %>% round(1)),
         bmi = paste(bmi_mean %>% round(1) , "Â±", bmi_sd %>% round(1)),
         blood_ige = paste(blood_ige_mean %>% round(1) , "Â±", blood_ige_sd %>% round(1)),
         `Blood_Ige > 150` = round(ige_150h / no_subject, 2) * 100,
         easi = paste(easi_mean %>% round(1) , "Â±", easi_sd %>% round(1)),
         scorad = paste(scorad_mean %>% round(1) , "Â±", scorad_sd %>% round(1)),
         oscorad = paste(oscorad_mean %>% round(1) , "Â±", oscorad_sd %>% round(1))) %>%
  select(group, `Female%`:oscorad) %>%
  column_to_rownames("group") %>% mutate_all(as.character) %>% 
  add_row(tibble_row(p_value)) %>% 
  t()

colnames(clinic_summary_t)[3] <- "p-value"
clinic_summary_t[6:8, 2:3] <- " / "
rownames(clinic_summary_t)[1] <- "Female(%)"
rownames(clinic_summary_t)[2] <- "Age"
rownames(clinic_summary_t)[3] <- "BMI"
rownames(clinic_summary_t)[4] <- "Blood IgE"
rownames(clinic_summary_t)[5] <- "Blood IgE>150(%)"
rownames(clinic_summary_t)[6] <- "EASI"
rownames(clinic_summary_t)[7] <- "SCORAD"
rownames(clinic_summary_t)[8] <- "oSCORAD"
colnames(clinic_summary_t)[2] <- "HC (n=30)"
knitr::kable(clinic_summary_t, 
             format = "html", 
             caption = "Baseline characteristics of AD and HC group") %>% 
  kableExtra::add_footnote("Age, bmi, and blood IgE were tested by t test. Gender and blood IgE > 150 (%) were tested by Fisher's exact test.")

clinic_summary_t_xlsx <- clinic_summary_t %>% as_tibble(rownames = " ")
clinic_summary_t_xlsx[9,1] <- "notes: Age, bmi, and blood IgE were tested by t test. Gender and blood IgE > 200 (%) were tested by Fisher's exact test."

```

```{r output table S1, eval=FALSE}
table_s1 <- extensive_meta %>% 
  filter(visit == "01") %>% 
  select(subject, gender, 
         visit, visit_quarter, age, bmi, 
         smoker, 
         drinks_per_week, 
         prick_test_result,
         skin_type_fitzpatrick_scale,
         age_onset, 
         other_atopic_disease = other_atopy, 
         family_atopic_dermatitis = family_eczema, 
         family_atopic_disease = family_atopy, 
         education,
         animal, 
         contains("scorad"), contains("easi")) %>% 
  mutate(age = age %>% round(1),
         bmi = bmi %>% round(2)) %>% 
  arrange(subject, visit) %>% distinct()
openxlsx::write.xlsx(table_s1, "data/supplementary/table_s1.xlsx", overwrite = T)
```

## Differential expressed genes: baseline (Table S2)

```{r DGE-V1, cache=TRUE, eval=FALSE}
# V1 LSvsNL --------------------
se_v1_LSvsNL <- se[, se$visit == "01" & se$skin_type %in% c("LS", "NL")]
de_v1_LSvsNL <- DESeq2::DESeqDataSet(se_v1_LSvsNL, ~ subject + skin_type)
de_v1_LSvsNL <- DESeq2::DESeq(de_v1_LSvsNL, parallel = TRUE)
de_v1_LSvsNL_LFC <- lfcShrink(de_v1_LSvsNL, coef = "skin_type_LS_vs_NL", type = "apeglm", parallel = TRUE, quiet = TRUE)
de_v1_LSvsNL_result <- de_v1_LSvsNL_LFC %>% as.data.frame() %>% tibble::rownames_to_column(var = "gene_name") %>% tibble()
de_v1_LSvsNL_result_s <- de_v1_LSvsNL_result %>% filter(padj < .05, abs(log2FoldChange) > 1)

# V1 LSvsNN --------------------
se_v1_LSvsNN <- se[, se$visit == "01" & se$skin_type %in% c("LS", "HC")]
de_v1_LSvsNN <- DESeq2::DESeqDataSet(se_v1_LSvsNN, ~ gender + skin_type) 
de_v1_LSvsNN <- DESeq2::DESeq(de_v1_LSvsNN, parallel = TRUE)
de_v1_LSvsNN_LFC <- lfcShrink(de_v1_LSvsNN, coef = "skin_type_LS_vs_HC", type = "apeglm", parallel = TRUE, quiet = TRUE)
de_v1_LSvsNN_result <- de_v1_LSvsNN_LFC %>% as.data.frame() %>% tibble::rownames_to_column(var = "gene_name") %>% tibble()
de_v1_LSvsNN_result_s <- de_v1_LSvsNN_result %>% filter(padj < .05, abs(log2FoldChange) > 1)

# V1 NLvsNN --------------------
se_v1_NLvsNN <- se[, se$visit == "01" & se$skin_type %in% c("NL", "HC")]
de_v1_NLvsNN <- DESeq2::DESeqDataSet(se_v1_NLvsNN, ~ gender + skin_type) 
de_v1_NLvsNN <- DESeq2::DESeq(de_v1_NLvsNN, parallel = TRUE)
de_v1_NLvsNN_LFC <- lfcShrink(de_v1_NLvsNN, coef = "skin_type_NL_vs_HC", type = "apeglm", parallel = TRUE, quiet = TRUE)
de_v1_NLvsNN_result <- de_v1_NLvsNN_LFC %>% as.data.frame() %>% tibble::rownames_to_column(var = "gene_name") %>% tibble()
de_v1_NLvsNN_result_s <- de_v1_NLvsNN_result %>% filter(padj < .05, abs(log2FoldChange) > 1)

# Consolidate the results -----
de_v1_c <-
  bind_rows(de_v1_LSvsNL_result_s %>% mutate(contrast = "LSvsNL"),
            de_v1_LSvsNN_result_s %>% mutate(contrast = "LSvsHC"),
            de_v1_NLvsNN_result_s %>% mutate(contrast = "NLvsHC")) %>% 
  mutate(direction = ifelse(log2FoldChange > 0, "up", "down"))
de_v1_c_s <-
  de_v1_c %>% 
  group_by(contrast) %>% 
  summarise(n_dys = n())
# No of dysregulated genes
n_de_v1_LSvsNL <- de_v1_c_s %>% filter(contrast == "LSvsNL") %>% pull(n_dys)
n_de_v1_LSvsNN <- de_v1_c_s %>% filter(contrast == "LSvsHC") %>% pull(n_dys)
n_de_v1_NLvsNN <- de_v1_c_s %>% filter(contrast == "NLvsHC") %>% pull(n_dys)
de_v1_full <- list(de_v1_LSvsNL = de_v1_LSvsNL_result, 
                   de_v1_LSvsNN = de_v1_LSvsNN_result, 
                   de_v1_NLvsNN = de_v1_NLvsNN_result)
de_v1_full %>% readr::write_rds("data/de_v1_full.rds")

```

```{r output table s2, eval=FALSE}
de_v1_c %>% readr::write_csv("data/supplementary/table_s2.csv")
```

```{r AD disease signature}
de_v1_c %>% 
  select(gene_name, log2FoldChange, padj, contrast) %>% 
  DT::datatable(rownames = F, options = list(
    order = list(list(1, "desc"), list(2, "aes"))
  )) %>% 
  DT::formatRound("log2FoldChange") %>% 
  DT::formatSignif("padj")
```

### Compare DEG list with GSE121212, and MADAD (new figure 1b)

```{r only protein coding genes}
clean_g_biomart <- function(gene_to_clean,
                            type = "hgnc_symbol"){
  g <- gene_to_clean
  mart <- biomaRt::useMart("ENSEMBL_MART_ENSEMBL")
  mart <- biomaRt::useDataset("hsapiens_gene_ensembl", mart)
  annotLookup <- biomaRt::getBM(
    mart = mart,
    attributes = c(
      "hgnc_symbol",
      "entrezgene_id",
      "ensembl_gene_id",
      "gene_biotype"),
    filter = type,
    values = g,
    uniqueRows=TRUE)
  g_p <- annotLookup %>% filter(gene_biotype == "protein_coding") %>% pull(hgnc_symbol)
  return(g_p)
}
```

```{r}
GENAD_DEG <- # GENAD DEG list
  readr::read_csv("data/supplementary/table_s2.csv")

GENAD_DEG_up <- 
  GENAD_DEG %>% 
  filter(log2FoldChange > 0) %>% 
  pull(gene_name) %>%
  unique %>% 
  clean_g_biomart()
# clipr::write_clip(GENAD_DEG_up)

GSE121212_DEG <-
  readRDS(url("https://rawcdn.githack.com/tuhulab/gse121212/ab5db865c53d44ee381294bbaec34c31013e1f52/data/rds/aLS_CO_DGE.rds")) %>% 
  bind_rows(readRDS(url("https://rawcdn.githack.com/tuhulab/gse121212/ab5db865c53d44ee381294bbaec34c31013e1f52/data/rds/aLS_NL_DGE.rds")))

GSE121212_DEG_up_clean <- 
  GSE121212_DEG %>% 
  filter(log2FoldChange > 0) %>% pull(gene_id) %>% 
  unique() %>% 
  clean_g_biomart()
# clipr::write_clip(GSE121212_DEG_up_clean)

MADAD <- 
  openxlsx::read.xlsx("~/Downloads/MADAD_original.xlsx", startRow = 2)
MADAD_up <- 
  MADAD %>% 
  filter(lgFCH_LSvsNL_estimate > 0) %>% 
  pull(ENTREZID) %>% 
  clean_g_biomart(type = "entrezgene_id")
# clipr::write_clip(MADAD_up)

# downregulated
GENAD_DEG_down <- 
  GENAD_DEG %>% 
  filter(log2FoldChange < 0) %>% 
  pull(gene_name) %>%
  unique %>% 
  clean_g_biomart()
# clipr::write_clip(GENAD_DEG_down)

GSE121212_DEG_down <- 
  GSE121212_DEG %>% 
  filter(log2FoldChange < 0) %>% pull(gene_id) %>% 
  unique() %>% 
  clean_g_biomart()
# clipr::write_clip(GSE121212_DEG_down)

MADAD_down <- 
  MADAD %>% 
  filter(lgFCH_LSvsNL_estimate < 0) %>% 
  pull(ENTREZID) %>% 
  clean_g_biomart(type = "entrezgene_id")
# clipr::write_clip(MADAD_down)

```

```{r}
library(gplots)

DEG_up <- list(
  GENAD = GENAD_DEG_up %>% unique,
  MADAD = MADAD_up %>% unique,
  GSE121212 = GSE121212_DEG_up_clean %>% unique
)

DEG_down <- list(
  GENAD = GENAD_DEG_down%>% unique,
  MADAD = MADAD_down%>% unique,
  GSE121212 = GSE121212_DEG_down%>% unique
)

DEG_up_venn <- 
  venn(DEG_up, show.plot = F)

DEG_down_venn <- 
  venn(DEG_down, show.plot = F)

de_v1_full <- 
  read_rds("data/de_v1_full.rds")$de_v1_LSvsNL

attributes(DEG_up_venn)$intersections %>% 
  enframe(name = "Set", value = "gene") %>% 
  mutate(direction = "up") %>% 
  bind_rows(
  attributes(DEG_down_venn)$intersections %>% 
  enframe(name = "Set", value = "gene") %>% 
  mutate(direction = "down")
  ) %>% 
  unnest() %>% 
  left_join(de_v1_full, by = c("gene" = "gene_name")) %>% 
  mutate(rank = -log10(padj) * log2FoldChange) %>% 
  group_by(Set, direction) %>% 
  summarise(`Median rank` = median(rank, na.rm = T) %>% round(2)) %>% 
  arrange(-`Median rank`) %>%
  pivot_wider(names_from = direction, values_from = `Median rank`) %>% 
  rename(`Median Rank (upregulated genes)` = up,
         `Median Rank (downregulated genes)` = down) %>% 
  write_tsv("data/_TMP/overlapping_deg_rank.txt")

  mutate(gene = map_chr(gene, ~ .x %>% str_c(collapse = " ")))

```


## Transcriptomic profile (Figure S1, Figure 1b, old Figure 1c)

### Intra-group cosine similarity (Figure S1a)

```{r fig-s1a-distance-Tissue type heterogenity, cache=TRUE, fig.cap="Cosine distance of three tissue types"}
cosine_st <-
    lapply(c("LS", "NL", "HC"), function(x){
    se %>% 
      subset(select = skin_type == x) %>% 
      assay(2) %>% 
      coop::cosine() %>% 
      as.numeric()})

names(cosine_st) <- c("LS", "NL", "HC")

cosine_df <- plyr::ldply(cosine_st, tibble, .name_repair = ~ c("cos_dist"),
                       .id = "skin_type") %>% tibble

cosine_df %>% 
  ggpubr::ggdensity(x = "cos_dist", fill = "skin_type", 
                    palette = pca_color, xlab = "Cosine similarity") %>% 
  ggpubr::ggpar(legend.title = "tissue type")
```

```{r median cosine similarity}
lapply(cosine_st, median)
```


### Heatmap for top genes (new Figure 1c)
```{r fig-s1b-heatmap-top100}
se <- 
  readRDS("data/se.rds")

top100_g <- 
  openxlsx::read.xlsx("data/_TMP/heatmap_g.xlsx")

se_heatmap_top100 <- se[top100_g$gene_name,] %>% 
  mutate(biopsy_area = ifelse(biopsy_area == "back_of_knee",
                              "back of knee",
                              biopsy_area))

rownames(se_heatmap_top100) <- top100_g$display_name

heatmap_data_m_top100 <-
  scale(assay(se_heatmap_top100, 2) %>% t() %>% log1p(), 
        center = TRUE, scale = TRUE) %>% t()

sample_site_color <- 
  colData(se_heatmap_top100) %>% as_tibble() %>% pull(biopsy_area) %>% unique()

color <- RColorBrewer::brewer.pal(length(sample_site_color), "Pastel2")
names(color) <- sample_site_color

gene_theme_color <- RColorBrewer::brewer.pal(length(top100_g$theme), "Accent")
names(gene_theme_color) <- top100_g$theme %>% unique()

heatmap_annotation <- 
  HeatmapAnnotation(
    `Tissue type` = colData(se_heatmap_top100) %>% as_tibble() %>% pull(skin_type), 
    `Anatomic region` = colData(se_heatmap_top100) %>% as_tibble() %>% pull(biopsy_area),
    col = list(`Tissue type` = 
                 c("LS" = "#eb2d0c", "NL" = "#eb8b9b", "HC" = "#91cf60"),
               `Anatomic region` = color),
    annotation_legend_param = 
      list(`Tissue type` = list(nrow = 1, title_gp = gpar(fontsize = 12),
                                labels_gp = gpar(fontsize = 10)),
           `Anatomic region` = list(nrow = 1, title_gp = gpar(fontsize = 12),
                                    labels_gp = gpar(fontsize = 10))
                                   ))

heatmap_top100 <- ComplexHeatmap::Heatmap(heatmap_data_m_top100,
                        name = "gene expression",
                        column_dend_reorder = TRUE,
                        # clustering_method_columns = "complete",
                        show_column_names = FALSE,
                        show_row_names = TRUE,
                        show_row_dend = FALSE,
                        row_names_gp = gpar(fontsize = 9),
                        top_annotation = heatmap_annotation,
                        heatmap_legend_param = list(
                          title_gp = gpar(fontsize = 12),
                          labels_gp = gpar(fontsize = 10)
                          ),
                        column_split = se_heatmap_top100$skin_type,
                        row_split = top100_g$theme,
                        row_title = gt_render(
                          c("Noncoding<br>RNA",
                            "Skin<br>barrier<br>and<br>physiology",
                            "Immune<br>response",
                            "G protein<br>coupled<br>receptor")
                        ),
                        row_title_rot = 0, 
                        row_title_side = "right",
                        row_title_gp = gpar(fill = gene_theme_color))

draw(heatmap_top100, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)
```

```{r output fig s1b, eval=FALSE}
png("data/supplementary/figure_1c.png", width = 480*6, height = 480*6, res=300)
draw(heatmap_top100, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)
dev.off()
```

### 3D-PCA plot for all DGEs (Figure 1b)
```{r figure 1b-pca}
core_genes <- 
  de_v1_c %>% 
  pull(gene_name) %>% 
  unique()

# dimension reduction
PCA_core <- 
  se[core_genes,] %>% 
  reduce_dimensions(method = "PCA")
PCA_core_d <- 
  attr(PCA_core, "internals")$PCA
```

```{r figure 1b-pca-vis, eval=FALSE}
pca3d::pca3d(PCA_core_d, 
             group = PCA_core$skin_type, 
             col = pca_color, radius = 1.5, 
             shape = "s", 
             show.plane = FALSE)

# save figure
rgl::snapshot3d("data/supplementary/figure_1b.png", fmt="png")
```

```{r figure 1b-pca-vis-show-vis, fig.cap = "pca 3d plot", dpi=300, out.width='80%'}
knitr::include_graphics("data/supplementary/figure_1b.png")
```

### Heatmap for all DGEs (Figure S1b)
```{r figure 1b}
se_heatmap <- se[core_genes,] %>% 
  mutate(biopsy_area = ifelse(biopsy_area == "back_of_knee",
                              "back of knee",
                              biopsy_area))
heatmap_data_m <-
  scale(assay(se_heatmap, 2) %>% t() %>% log1p(), 
        center = TRUE, scale = TRUE) %>% t()
sample_site_color <- 
  colData(se_heatmap) %>% as_tibble() %>% pull(biopsy_area) %>% unique()
color <- RColorBrewer::brewer.pal(length(sample_site_color), "Pastel2")
names(color) <- sample_site_color
heatmap_annotation <- 
  HeatmapAnnotation(
    `Tissue type` = colData(se_heatmap) %>% as_tibble() %>% pull(skin_type), 
    `Anatomic region` = colData(se_heatmap) %>% as_tibble() %>% pull(biopsy_area),
    col = list(`Tissue type` = pca_color,
               `Anatomic region` = color),
    annotation_legend_param = list(`Tissue type` = list(nrow = 1, 
                                                        title_gp = gpar(fontsize = 16),
                                                        labels_gp = gpar(fontsize = 16)),
                                   `Anatomic region` = list(nrow = 1,
                                                            title_gp = gpar(fontsize = 16),
                                                            labels_gp = gpar(fontsize = 16))
                                   ))

heatmap <- ComplexHeatmap::Heatmap(heatmap_data_m,
                        name = "gene expression",
                        column_dend_reorder = TRUE,
                        # clustering_method_columns = "complete",
                        show_column_names = FALSE,
                        show_row_names = FALSE,
                        show_row_dend = FALSE,
                        # row_names_gp = gpar(fontsize = min(10, 800 * 1.5 / dim(heatmap_data_m)[1])),
                        top_annotation = heatmap_annotation,
                        heatmap_legend_param = list(
                          title_gp = gpar(fontsize = 16),
                          labels_gp = gpar(fontsize = 14)
                          ))

draw(heatmap, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)
```

```{r output figure 1b, eval=FALSE}
png("data/supplementary/figure_1c.png", width = 480*6, height = 480*6, res=300)
draw(heatmap, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)
dev.off()
```

## Gene set enrichment analysis (Figure 2, Table S3)
```{r dge result}
gse107361 <- 
  readr::read_rds("data/dge_gse107361.rds") %>% # adult vs children
  mutate(dge_result = map(dge_result, unnest))
gse120721 <- readr::read_rds("data/dge_gse120721.rds") # LSM
gse121212 <- readr::read_rds("data/dge_gse121212.rds")
ChronicvsAcute <- readr::read_rds("data/dge_AcutevsChronic.rds")

dge_result <-
  tibble(
    project = c(rep("multi-omics",length(de_v1_full)), 
                rep("gse121212",nrow(gse121212)),
                rep("gse107361",nrow(gse107361)),
                rep("gse120721",nrow(gse120721)),
                rep("acute_chronic", nrow(ChronicvsAcute))),
    contrast = c(de_v1_full %>% names %>% str_remove("de_v1_"), 
                 gse121212$contrast, 
                 gse107361$contrast,
                 gse120721$contrast,
                 ChronicvsAcute$contrast),
    dge_result = c(de_v1_full, 
                   gse121212$res_shrink,
                   gse107361$dge_result,
                   gse120721$dge_result,
                   ChronicvsAcute$data))
```

```{r prepare gene set, cache=TRUE}
# prepare gs --------------
gs <- 
  tibble(gs_name = c("CP:REACTOME", 
                     "TFT:GTRD", 
                     "GO:BP", 
                     "GO:MF")) %>% 
  mutate(gs_list = map(gs_name, function(subcat){
  gs_df <- msigdbr::msigdbr(species = "Homo sapiens", 
                            subcategory = subcat) %>% 
    group_by(gs_name) %>% tidyr::nest() %>% 
    mutate(gene_id = purrr::map(data, ~ .x %>% pull(gene_symbol))) %>% select(-data)
  gs_list <- gs_df$gene_id
  names(gs_list) <- gs_df$gs_name
  return(gs_list)
}))
```

```{r fgsea, cache=TRUE}
gsea_exp <- 
  expand.grid(contrast = dge_result$contrast,
              gs_name = gs$gs_name) %>% 
  left_join(dge_result, by = "contrast") %>% 
  left_join(gs, by = "gs_name") %>% 
  mutate(rank = map2(dge_result, gs_name,
                     function(dge_result, gs_name){
                       p_cutoff <- ifelse(gs_name %in% c("MIR:MIRDB", 
                                                         "TFT:GTRD"),
                                          1, # .05
                                          1)
                       log2fc_cutoff <- ifelse(gs_name %in% c("MIR:MIRDB",
                                                              "TFT:GTRD"),
                                               0, # 1
                                               0)
                       dge_result %>% 
                         mutate(rank = log2FoldChange) %>% 
                         filter(padj < p_cutoff, abs(log2FoldChange) > log2fc_cutoff) %>% 
                         select(gene_name, rank) %>% deframe}),
         gse_res = map2(rank, gs_list,
                            function(rank, gs_list){
                              gsa_stat <- fgsea::fgsea(pathways = gs_list, 
                                                       stats    = rank,
                                                       minSize  = 8,
                                                       maxSize  = 500, 
                                                       eps = 0,
                                                       nPermSimple = 10000) %>% 
                            as_tibble()
}))

gsea_res <- gsea_exp %>% select(contrast, gs_name, project, gse_res) %>% unnest()
gsea_res %>% saveRDS("data/gsea_res_FIVE_nocutoff.rds")
```

```{r gsea split}
gsea_res <- 
  gsea_res %>% 
  mutate(contrast = contrast %>% str_remove("de_v1_")) %>% 
  filter(gs_name %in% c("CP:REACTOME", 
                        # "MIR:MIRDB", 
                        "TFT:GTRD", 
                        "GO:BP", 
                        "GO:MF"))
```

### Table S3

```{r appendix}
gsea_res_ex_filt <- 
  gsea_res %>% filter(padj < .05, !contrast %>% stringr::str_detect("PSO"))

gsea_res_ex_n <- 
  gsea_res_ex_filt %>% 
  group_by(pathway) %>% 
  summarise(n = n())

gsea_res_ex_appendix_w <-
  gsea_res_ex_filt %>% mutate(contrast = paste(project, contrast)) %>% 
  select(contrast, pathway, NES) %>% pivot_wider(names_from = contrast, values_from = NES) %>% 
  left_join(gsea_res_ex_n, by = "pathway") %>% arrange(-n) %>% 
  select(pathway, 
         n,
         contains("multi-omics"), 
         contains("gse121212"), 
         contains("acute_chronic"),
         contains("gse107361"), 
         contains("gse120721"))
# saveRDS(gsea_res_ex_appendix_w, "data/gsea_appendix.rds")

gsea_res_ex_filt_w <-
  gsea_res_ex_filt %>% mutate(contrast = paste(project, contrast)) %>% 
  group_by(gs_name) %>% nest() %>% 
  mutate(data = map(data, 
                    ~ .x %>% 
                      select(contrast, pathway, NES) %>% 
                      pivot_wider(names_from = contrast, values_from = NES) %>% 
                      left_join(gsea_res_ex_n, by = "pathway") %>% 
                      arrange(-n) %>% 
                      mutate(url = paste0("https://www.gsea-msigdb.org/gsea/msigdb/cards/",
                                          pathway,
                                          ".html")) %>% 
  select(pathway,
         n,
         url,
         contains("multi-omics"), 
         contains("gse121212"), 
         contains("acute_chronic"),
         contains("gse107361"), 
         contains("gse120721")) 
  ))

gsea_res_ex_filt_p_value_w <-
  gsea_res_ex_filt %>% mutate(contrast = paste(project, contrast)) %>% 
  group_by(gs_name) %>% nest() %>% 
  mutate(data = map(data, 
                    ~ .x %>% 
                      select(contrast, pathway, padj) %>% 
                      pivot_wider(names_from = contrast, values_from = padj) %>% 
                      left_join(gsea_res_ex_n, by = "pathway") %>% 
                      arrange(-n) %>% 
                      mutate(url = paste0("https://www.gsea-msigdb.org/gsea/msigdb/cards/",
                                          pathway,
                                          ".html")) %>% 
  select(pathway,
         n,
         url,
         contains("multi-omics"), 
         contains("gse121212"), 
         contains("acute_chronic"),
         contains("gse107361"), 
         contains("gse120721")) 
  ))


gsea_xlsx <- c(gsea_res_ex_filt_w$data,
                  gsea_res_ex_filt_p_value_w$data)

names(gsea_xlsx) <- 
  c(gsea_res_ex_filt_w$gs_name %>% str_replace(":", "_") %>% str_c("_NES"),
    gsea_res_ex_filt_w$gs_name %>% str_replace(":", "_") %>% str_c("_padj"))

gsea_xlsx %>% openxlsx::write.xlsx("data/supplementary/table_s3_gsea_res.xlsx", overwrite = T)
```

### Figure 2
```{r gsea-vis}
# gene set ad related
gs_ad <- gsea_res_ex_filt %>% filter(pathway %in% c("REACTOME_CHEMOKINE_RECEPTORS_BIND_CHEMOKINES",
                                                    "REACTOME_INTERFERON_SIGNALING",
                                                    "KEGG_JAK_STAT_SIGNALING_PATHWAY",
                                                    "REACTOME_INTERLEUKIN_10_SIGNALING",
                                                    "GOBP_INFLAMMATORY_RESPONSE",
                                                    "REACTOME_INTERLEUKIN_4_AND_INTERLEUKIN_13_SIGNALING", # IL4&13
                                                    "REACTOME_INTERLEUKIN_12_FAMILY_SIGNALING", # IL12
                                                    "REACTOME_INTERLEUKIN_1_FAMILY_SIGNALING", # IL1
                                                    "REACTOME_INTERLEUKIN_10_SIGNALING",
                                                    "REACTOME_INTERLEUKIN_20_FAMILY_SIGNALING",
                                                    # "REACTOME_INTERLEUKIN_17_SIGNALING",
                                                    "GOBP_T_CELL_ACTIVATION",
                                                    "GOBP_T_CELL_PROLIFERATION",
                                                    "GOBP_KERATINOCYTE_DIFFERENTIATION",
                                                    "GOBP_NEUTRAL_LIPID_BIOSYNTHETIC_PROCESS",
                                                    "GOBP_CHRONIC_INFLAMMATORY_RESPONSE",
                                                    "GOBP_POSITIVE_REGULATION_OF_MACROPHAGE_ACTIVATION",
                                                    "GOBP_EOSINOPHIL_MIGRATION",
                                                    "GOBP_DEFENSE_RESPONSE_TO_BACTERIUM") & padj < .05
                               # pathway %>% str_detect("REACTOME_INTERLEUKIN") & padj < .05
                             ) %>% pull(pathway) %>% unique()

gs_r <- 
  gsea_res %>% 
  filter(
    (gs_name %in% c("TFT:GTRD") & padj < .05 & 
       project == "multi-omics")) %>% pull(pathway) %>% unique()

gsea_res_filt <-
  gsea_res %>% filter(pathway %in% c(gs_r, gs_ad)) %>% mutate(`-log10p` = -log10(padj)) %>% 
  group_by(pathway, gs_name) %>% nest() %>% 
  mutate(url = paste0("https://www.gsea-msigdb.org/gsea/msigdb/cards/",
                      pathway,
                      ".html"),
         des = map_chr(url, 
                   function(url){
                     url %>% xml2::read_html() %>% 
                       rvest::html_node(css = "tr:nth-child(3) td") %>%
                       rvest::html_text()}),
         des = ifelse(gs_name == "GO:BP", # add description
                      pathway %>% str_remove("GOBP") %>% str_replace_all("_", " ") %>% str_to_sentence(), 
                      des),
         des = ifelse(gs_name %in% c("MIR:MIRDB", "TFT:GTRD"),
                      pathway %>% str_replace("_TARGET_GENES", " target genes"),
                      des)) %>% 
  unnest() %>% 
  mutate(contrast = case_when(
           contrast == "AD-LSvsAD-NL" ~ "LSvsNL",
           contrast == "AD-LSvsCO" ~ "LSvsHC",
           contrast == "AD-NLvsCO" ~ "NLvsHC",
           contrast == "LSvsNN" ~ "LSvsHC",
           contrast == "NLvsNN" ~ "NLvsHC",
           contrast == "Adult_LSvsAdult_NL" ~ "Adult_LSvsNL",
           contrast == "Adult_NNvsChildren_NN" ~ "Adult_HCvsChildren_HC",
           contrast == "Children_LSvsChildren_NL" ~ "Children_LSvsNL",
           contrast == "LS/NL_derm" ~ "LSvsNL_dermis",
           contrast == "LS/NL_epi" ~ "LSvsNL_epidermis",
           contrast == "LS/NL_FT" ~ "LSvsNL_full tissue",
           TRUE ~ contrast)) 


gsea_ad_p <-
  gsea_res_filt %>% filter(pathway %in% gs_ad, 
                           !contrast %>% str_detect("PSO"),
                           padj < .05,
                           !contrast %in% c("Adult_LSvsAdult_NN",
                                            "Adult_NLvsAdult_NN",
                                            "Adult_NLvsAdult_NN",
                                            "Children_NLvsChildren_NN",
                                            "Children_LSvsChildren_NN",
                                            "Adult_LSvsChildren_LS",
                                            "Adult_NLvsChildren_NL",
                                            "Adult_HCvsChildren_HC",
                                            "LS/NN_derm",
                                            "LS/NN_epi",
                                            "LS/NN_FT",
                                            "NL/NN_derm",
                                            "NL/NN_FT")) %>% 
  mutate(
    project = case_when(project == "multi-omics" ~ "GENAD",
                        project == "acute_chronic" ~ "Acute vs. Chronic",
                        TRUE ~ toupper(project)),
    # project = ifelse(project == "multi-omics", "GENAD", project),
    project = factor(project, levels = c("GSE121212",
                                         "Acute vs. Chronic",
                                         "GSE107361",
                                         "GSE120721",
                                         "GENAD")),
    des = des %>% str_replace_all("interleukin|Interleukin", "IL"),
    cat = case_when(
    pathway %in% c("PID_IL23_PATHWAY",
                   "KEGG_JAK_STAT_SIGNALING_PATHWAY",
                   "REACTOME_INTERLEUKIN_10_SIGNALING",
                   "REACTOME_INTERLEUKIN_4_AND_INTERLEUKIN_13_SIGNALING",
                   "REACTOME_INTERLEUKIN_12_FAMILY_SIGNALING",
                   "WP_TYPE_II_INTERFERON_SIGNALING_IFNG",
                   "KEGG_JAK_STAT_SIGNALING_PATHWAY") ~ "signal transduction",
    pathway %in% c("GOBP_KERATINOCYTE_DIFFERENTIATION",
                   "GOBP_NEUTRAL_LIPID_BIOSYNTHETIC_PROCESS", "REACTOME_DEFENSINS",
                   "GOBP_DEFENSE_RESPONSE_TO_BACTERIUM") ~ "skin barrier",
    pathway %in% c("KEGG_CHEMOKINE_SIGNALING_PATHWAY",
                   "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION",
                   "GOBP_INFLAMMATORY_RESPONSE",
                   "GOBP_T_CELL_ACTIVATION",
                   "GOBP_T_CELL_PROLIFERATION",
                   "GOBP_CHRONIC_INFLAMMATORY_RESPONSE") ~ "immune response",
    TRUE ~ "signal transduction")) %>% 
  ggplot(aes(contrast, des, size = `-log10p`, color = NES)) + 
  geom_point() +
  scale_size_area(max_size = 8) + 
  scale_color_gradient2(low="blue", mid = "white", high="red") +
  theme_bw() +
  facet_grid(cat~project, scales = "free", space = "free") +
  theme(axis.title = element_blank(), axis.text.x = element_text(angle = 35, hjust = 1))

gsea_ad_p
```

```{r output figure 2, eval=FALSE}
ggsave("data/supplementary/figure_2.png", gsea_ad_p, width = 10.5, height = 7)
```


## Variance parition analysis (Figure 3, Table S4, Table S5)

```{r VP core gene, cache=TRUE, dpi=600, fig.width=12}
# transform data for linear modeling
se <- read_rds("data/se.rds")
core_genes <- 
  read_csv("data/supplementary/table_s2.csv") %>%
  pull(gene_name) %>% unique

se_core <- se[intersect(rownames(se), core_genes), 
              which(se$rna_quality %in% c("high"))]

seExpr_core <- calcNormFactors(se_core)
design_core <- model.matrix(~ skin_type, seExpr_core$samples)
vobj_core <- voom(seExpr_core, design_core)

# perform VP analysis
vp_core <- variancePartition::fitExtractVarPartModel(
    vobj_core,
    ~ 
      scorad + easi_total_score +
      (1|skin_type) + (1|subject) + (1|visit_quarter) + 
      (1|gender) + (1|biopsy_area),
    colData(se_core) %>% as_tibble()
)
vp_core_t <- 
  vp_core %>% as_tibble(rownames = "gene_name")

vp_core_t %>% saveRDS("data/vp_core_t.rds")

# plot 1 - violin
vp_core_p <- vp_core
names(vp_core_p)[1] <- "anatomic region"
names(vp_core_p)[3] <- "tissue type (LS/NL/HC)"
names(vp_core_p)[4] <- "individual"
names(vp_core_p)[5] <- "quarter"
names(vp_core_p)[6] <- "SCORAD"
names(vp_core_p)[7] <- "EASI"
names(vp_core_p)[8] <- "residual"
source("R/plot.R")
vpcore_violin <-
  plotVarPart_internal(vp_core_p) + 
  theme(plot.margin = margin(.5, .5, .5, 2, "cm")) +
  ylab("Var(%) explained")

# plot 2 - cytokine dot plot
cytokine_vp <- 
  vp_core_t %>% 
  filter(gene_name %>% str_detect("^IL\\d{1}|TNF|EBI|IFN|OSMR"),
         gene_name != "IL4I1") %>% 
  arrange(-skin_type) %>% 
  pivot_longer(cols = !gene_name, names_to = "term") %>% 
  filter(term == "skin_type", value > .05) %>% 
  mutate(var_explained = round(value * 100, 2))
cytokine_vp_P <- ggpubr::ggdotchart(
  cytokine_vp, x = "gene_name", y = "var_explained",
  color = "#FC4E07",
  xlab = "",
  sorting = "descending",
  rotate = TRUE, 
  add = "segments",
  ggtheme = ggpubr::theme_pubr(),
  dot.size = 4,
  ylab = "Var% explained by tissue type (LS/NL/HC)"
)

# plot 3 - top cytokine boxplot
top_cytokine_FVE <-
  c(
  cytokine_vp %>% filter(term == "skin_type") %>% 
  arrange(-var_explained) %>% slice_head(n = 5) %>% pull(gene_name) %>% 
  discard(function(x){x == "IL4I1"}),
  "UGT3A2", "TTC38", "CHRM4", "LINC01605", "LINC02345"
  )

top_cytokine_boxplot <-
  se[top_cytokine_FVE,] %>% as_tibble() %>% 
  mutate(feature = .feature %>% forcats::fct_relevel(top_cytokine)) %>% 
  ggboxplot(x = "skin_type", y = "counts_scaled", add = "jitter", 
            facet.by = "feature", xlab = FALSE,
            ylab = "counts", color = "skin_type", 
            palette = pca_color, legend.title = "",
            font.legend = c(14), nrow = 2, yscale = "log10")
  # stat_compare_means()

# consolidate 3 plots
vp_core_p <- 
  ((vpcore_violin / top_cytokine_boxplot) | cytokine_vp_P)
vp_core_p
```

### Figure 3
```{r output figure 3a, eval=FALSE}
ggsave("data/supplementary/figure_3.png", vp_core_p, 
       width = 12 * 1.2, height = 8 * 1.2, 
       dpi = "retina")
```

### Table S4
```{r output table s4, eval=FALSE}
vp_core_t %>% arrange(-skin_type) %>% 
  readr::write_csv("data/supplementary/table_s4.csv")
```

## Time effect - stepwise DGE analysis

```{r load and clean data}
se <- readr::read_rds("data/se.rds")
se <- se[,!is.na(se$visit_quarter)]
```

### Time stability (dge_tstat)

```{r dge_time_alone, eval=FALSE}
dge_tstat_design <-
  expand_grid(
  skin_type = colData(se)$skin_type %>% unique,
  t = c("visit", "visit_quarter")
  ) %>% 
  mutate(name = paste(skin_type, t, sep = "_"))
  
dge_tstat_result <-
  plyr::llply(1:nrow(dge_tstat_design), function(index){
    se <- se[,se$skin_type == dge_tstat_design$skin_type[index]]
    f <- paste("~", dge_tstat_design$t[index])
    deseq <- DESeqDataSet(se, design = as.formula(f))
    deseq <- DESeq(deseq, test = "LRT", reduced = ~ 1, 
                   parallel = T, 
                   BPPARAM = MulticoreParam(min(detectCores() - 2, 8)))
    deseq <- results(deseq)
    deseq <- as_tibble(deseq, rownames = "gene_name") %>% 
      mutate(name = dge_tstat_design$name[[index]])
  }, .progress = "text")
# saveRDS(dge_tstat_result, "data/dge_tstat_result.rds")
# dge_tstat_result <- readr::read_rds("data/dge_tstat_result.rds")

dge_tstat_result_filt <-
  map_dfr(dge_tstat_result,
           ~ filter(.x, padj < .05, 
                    abs(log2FoldChange) > 1)) 

library(enrichR)
setEnrichrSite("Enrichr")

dge_tstat_result_filt_n <- 
  dge_tstat_result_filt %>% group_by(name) %>% nest() %>% 
  mutate(gene_name = map(data, ~ pull(.x, gene_name)),
         enrichment = map(gene_name, ~ enrichr(., c("Reactome_2016"))))

dge_tstat_g <- dge_tstat_result_filt$gene_name %>% unique
dbs <- c("Human_Gene_Atlas", "ARCHS4_Tissues")
enriched_dge_tstat <- enrichr(dge_tstat_g, dbs)

adipocyte <- 
  map2_dfr(enriched_dge_tstat, dbs, 
         ~ .x %>% mutate(name = .y) %>% 
           filter(Adjusted.P.value < .05))
```

### adipocytes
```{r}
se <- readr::read_rds("data/se.rds")

se_adipocyte <- 
  se[c("CIDEC",    # CIDEC cell death inducing DFFA like effector c
       "FABP4",    # FABP4 fatty acid binding protein 4
       "PLIN4"),]  # perilipin 4

se_adipocyte %>% as_tibble() %>% 
  # group_by(sample, feature) %>% 
  # summarise(sum_count = sum(counts_scaled + .5) %>% log2()) %>% 
  ggplot(aes(log2(counts_scaled + .1))) + 
  facet_grid(feature~.) +
  geom_histogram()
```

```{r}
se_adipocyte %>% as_tibble() %>% 
  ggplot(aes(skin_type, counts_scaled)) + 
  facet_grid(~feature) +
  geom_boxplot()
```

```{r}
# Adipocyte differentiation marker
# Transient Receptor Potential Cation Channel Subfamily V
se[c("TRPV2", "TRPV1", "TRPV3", "TRPV4", "TRPV6"),] %>% 
  as_tibble() %>% 
    ggplot(aes(skin_type, counts_scaled)) + 
    facet_grid(~feature) +
    geom_boxplot()
```


### Disease signature stability
```{r load dge-time}
dge_time <- 
  readr::read_rds("data/dge_time.rds")

names(dge_time)[1:2] <- 
  c("DGE_condition_specific",
    "DGE_condition_specific_non")

dge_time[[2]] %>% unnest(cols = "result") %>% 
  filter(padj < .05, abs(log2FoldChange) > 1) %>% 
  group_by(t, contrast) %>% 
  summarise(n = n())

ad_sig_stat_g <- 
  dge_time[[2]] %>% unnest(cols = "result") %>% 
    filter(padj < .05, abs(log2FoldChange) > 1) %>% 
  filter(contrast =="LSvsNL", t =="visit_quarter") %>% 
  pull(gene_name)

```

### Condition(LS/NL/HC)-specific change
```{r}
dge_time[[1]] %>% unnest(cols = "result") %>% 
  filter(padj < .05, abs(log2FoldChange) > 1) %>% 
  group_by(t, contrast) %>% summarise(n = n())

condition_specific_g <- 
  dge_time[[1]] %>% unnest(cols = "result") %>% 
  filter(padj < .05, abs(log2FoldChange) > 1) %>% 
  group_by(t, contrast) %>% nest() 

g_tissue_injury <-
  se[c("SAA2", "MUC1", "MGST1", "PTGS2"),] %>% as_tibble() %>% 
    group_by(.feature, visit_quarter) %>% 
    mutate(group_mean = mean(counts_scaled),
           outlier = counts_scaled > 5 * group_mean,
           outlier = ifelse(outlier, .8, .3)
           # outlier = 0.05
           ) %>%
    ggplot(aes( x = visit_quarter, 
                y = log2(counts_scaled + 1),
                color = skin_type, 
                fill = skin_type)) +
    geom_jitter(aes(alpha = outlier) ,binaxis = "y", stackdir = "center") +
    # geom_dotplot(aes(alpha = outlier) ,binaxis = "y", stackdir = "center") +
    facet_wrap(~ .feature, scales = "free", nrow = 2) + 
    stat_summary(aes( group= skin_type), fun = mean, geom= "line", size = 2, alpha = .95) +
    scale_fill_manual(values = pca_color) +
    scale_color_manual(values = pca_color) +
    ylab("log2 counts") + 
    xlab("Quarter of visit") +
    theme_classic()

  
  
  
    # ggline(x = "visit_quarter", 
    #        y = "counts_scaled",
    #        facet.by = ".feature",
    #        add = c("mean", "dotplot"),
    #        color = "skin_type", 
    #        palette = pca_color,
    #        xlab = "Quarter of visit",
    #        ylab = "Counts") %>% 
    # ggpar(yscale = "log10")

g_subcutis <-
  se[c("CIDEC", "FABP4", "PLIN4"),] %>% as_tibble() %>% 
    group_by(.feature, visit_quarter) %>% 
    mutate(group_mean = mean(counts_scaled),
           outlier = counts_scaled > 5 * group_mean,
           outlier = ifelse(outlier, .8, .3)
           # outlier = 0.05
           ) %>%
    ggplot(aes( x = visit_quarter, 
                y = log2(counts_scaled + 1),
                color = skin_type, 
                fill = skin_type)) +
    geom_jitter(aes(alpha = outlier)) +
    # geom_dotplot(aes(alpha = outlier) ,binaxis = "y", stackdir = "center") +
    facet_wrap(~ .feature, scales = "free", nrow = 2) + 
    stat_summary(aes( group= skin_type), fun = mean, geom= "line", size = 2, alpha = .95) +
    scale_fill_manual(values = pca_color) +
    scale_color_manual(values = pca_color) +
    ylab("log2 counts") + 
    xlab("Quarter of visit") +
    theme_classic()

```

### Time outlier figure
```{r}
time_variation_t <- 
  dge_tstat_result_filt %>% 
  select(gene_name, name) %>% 
  group_by(name) %>% 
  nest %>% 
  mutate(d = map(data, ~ se[.x$gene_name, ] %>% 
                   as_tibble)) %>% 
   mutate(skin_type = name %>% str_extract("(LS|NL|NN)"),
          skin_type = ifelse(skin_type == "NN", "HC", skin_type),
          time_type = name %>% str_extract("(quarter|visit$)")) %>% 
  ungroup %>% 
  select(skin_type, time_type, d)
# saveRDS(time_variation_t, "data/_TMP/time_variation_t.rds")


time_outlier <-
  dge_tstat_result_filt %>% 
  select(gene_name, name) %>% 
  mutate(skin_type = str_extract(name, "(LS|NL|NN)"),
         skin_type = ifelse(skin_type == "NN", "HC", skin_type),
         t = str_extract(name, "(visit$|visit_quarter)")) %>% 
  select(-name) %>% 
  group_by(skin_type, t) %>% nest() %>% 
  mutate(g_n = map(data, "gene_name")) %>% 
  mutate(figure = 
           pmap(list(g_n, skin_type, t),
                       function(g_n, skin_type, t){
  plot <-
    se[g_n, se$skin_type == skin_type] %>% as_tibble() %>% 
    ggboxplot(x = t,
              y = "counts_scaled",
              add = "jitter",
              facet.by = ".feature",
              xlab = ifelse(t == "visit_quarter", 
                            "quarter",
                            t),
              ylab = FALSE,
              yscale = "log10",
              scales = "free", title = skin_type)
  }),
        save_f = pmap(list(skin_type, t, figure),
                      function(skin_type, t, figure){
          name = paste0("supplement_t_outlier_", paste(skin_type, t, sep = "_"),".png")
          ggsave(filename = paste0("figure/supplement_t_outlier", name),
                 figure, width = 16.8, height = 11.8)
        }))

```

```{r fluctuating}
time_variation_g <- 
  se[c("IL34", "IL37", "CCL22", "IL2RB", 
     "SERPINB4", "S100A9", "HRNR", 
     "COL4A4", "OSM", "IL36A", "IL4R", "KRT16"), ] %>% 
  as_tibble() %>% 
  group_by(.feature, skin_type, visit_quarter) %>% 
  ggplot(aes( x = visit_quarter, 
              y = log2(counts_scaled + 1),
              color = skin_type, 
              fill = skin_type)) +
  geom_dotplot(alpha = .1, binaxis = "y", stackdir = "center") +
  facet_wrap(~ .feature, scales = "free", nrow = 2) + 
  stat_summary(aes( group= skin_type), fun = mean, geom= "line", size = 2, alpha = .95) +
  scale_fill_manual(values = pca_color) +
  scale_color_manual(values = pca_color) +
  ylab("log2 counts") + 
  xlab("Quarter of visit") +
  theme_classic()

ggsave("data/supplementary/time_variation_g.png",
       time_variation_g,
       width = 10 , height = 4, 
       dpi = "retina")



g_subcutis
```


## Space effect - DGE anotomic area (Figure 4b 4c)
```{r DGE space, cache=TRUE, eval=FALSE}
DGE_space <-
  tibble(
    C1 = c("LS", "HC"),
    C2 = c("NL", "HC")
  ) %>% 
  mutate(
    contrast = paste0(C1, "vs", C2),
    se_exp = map2(C1, C2, ~ se[,se$skin_type %in% c(.x, .y)]),
    design_f = ifelse(C1 == "HC" & C2 == "HC", 
                      "~ subject + visit + biopsy_area", 
                      "~ subject + skin_type + visit + biopsy_area"),
    deseq = map2(se_exp, design_f, ~ DESeqDataSet(.x, .y %>% as.formula)),
    deseq = map(deseq, ~ DESeq(.x, parallel = T)),
    res_name = map(deseq, ~ resultsNames(.x) %>% grep(pattern = "biopsy_area", ., value = T))) %>% 
  unnest(cols = res_name) %>% 
  mutate(
    res_shrink = map2(deseq, res_name, ~ lfcShrink(.x, coef = .y, type = "apeglm", parallel = T)),
    res_shrink = map(res_shrink, ~ as_tibble(.x, rownames = "gene_name"))
  )
```

### Table S5
```{r output table s5, eval=FALSE}
DGE_space_filter <- 
  DGE_space %>% select(-se_exp, -design_f, -deseq) %>% 
  unnest(cols=res_shrink) %>% 
  filter(padj < .05, abs(log2FoldChange)>1)

readr::write_csv(DGE_space_filter, "data/supplementary/table_s5_dge_space.csv")
```

## Space effect - biological replicates

### Biological replicates
```{r found biological replicates}
bioreplicates <-
  colData(se) %>% as_tibble() %>% 
  mutate(biological_rep_id = paste(subject, visit, skin_type)) %>%
  group_by(biological_rep_id) %>% 
  summarise(n = n()) %>% 
  filter(n == 2) %>% 
  pull(biological_rep_id)

bioreplicates_t <- 
  colData(se) %>% as_tibble() %>% 
  mutate(biological_rep_id = paste(subject, visit, skin_type)) %>% 
  filter(biological_rep_id %in% bioreplicates) %>% 
  left_join(se[c("CCL22"),] %>% as_tibble())
# saveRDS(bioreplicates_t, "data/_TMP/bioreplicate_t.rds")

replicate_g <-
  bioreplicates_t %>% 
    select(biological_rep_id, 
           subject, visit, skin_type, feature, 
           counts_scaled, replicate_ID) %>% 
  ggplot(aes(x = replicate_ID, 
             y = counts_scaled, 
             group = biological_rep_id)) + 
  geom_point(aes(color = skin_type)) + 
  facet_grid(feature ~ skin_type, scales = "free") + 
  geom_line(aes(color = skin_type), alpha = .5) + 
  geom_boxplot(aes(group = replicate_ID, fill = skin_type), alpha = .5) + 
  scale_color_manual(values = pca_color) + 
  scale_fill_manual(values = pca_color) + 
  scale_y_continuous(trans = "log2") +
  xlab("replicate") +
  theme(axis.title.y = element_blank())

# ggsave("data/supplementary/supplement_replicate.png", replicate_g,
#        height = 25, width = 8)
```

```{r}
t <-
  readRDS(url("https://cdn.jsdelivr.net/gh/tuhulab/Shiny_AD_time_space@master/data/bioreplicate_t.rds")) %>%
  select(biological_rep_id,
         subject, visit, skin_type, feature,
         counts_scaled, replicate_ID) %>%
  arrange(biological_rep_id) %>%
  mutate(counts_scaled = counts_scaled + 1) %>%
  filter(feature %in% "CCL22") %>% 
  mutate(skin_type = 
           case_when(skin_type == "LS" ~ "Lesional",
                     skin_type == "NL" ~ "Non-lesional",
                     skin_type == "HC" ~ "Healthy control"),
         skin_type = 
           factor(skin_type, levels = c("Healthy control",
                                        "Non-lesional",
                                        "Lesional")))

stat_test <-
  t %>% group_by(feature, skin_type) %>%
  t_test(counts_scaled ~ replicate_ID, paired = T) %>%
  add_significance() %>%
  add_xy_position(x = "replicate_ID",
                  y.trans = log2, step.increase = .08) %>%
  group_by(feature) %>% mutate(y.position = max(y.position)) %>%
  ungroup()

g_bioreplicate <- 
t %>% filter(counts_scaled > 50) %>% 
    pivot_wider(names_from = replicate_ID, values_from = counts_scaled) %>% 
ggpaired(cond1 = "01", cond2 = "02", y = "counts_scaled",
         xlab = "Biological replicate",
         ylab = "",
         line.color = "grey", line.size = .4,
         fill = "skin_type",
         palette = c("Lesional" = "#eb2d0c", 
                     "Non-lesional" = "#eb8b9b", 
                     "Healthy control" = "#91cf60")) %>%
    facet(facet.by = c("skin_type"), scales = "free_y") +
    stat_pvalue_manual(stat_test, label = "p") +
    scale_y_continuous(trans = "log2", expand = expansion(mult = c(.05, .1)), breaks = c(10, 100, 1000, 10000, 30000)) +
  rremove("legend")


```


### correlation coefficient
```{r}
ad_sig <- 
  readr::read_csv("data/supplementary/table_s2.csv",
                  show_col_types = F) %>% 
  pull(gene_name) %>% 
  unique

bioreplicates_sig_t <- 
  colData(se) %>% as_tibble() %>% 
  mutate(biological_rep_id = paste(subject, visit, skin_type)) %>% 
  filter(biological_rep_id %in% bioreplicates) %>% 
  left_join(se[ad_sig,] %>% as_tibble())
```


## Space effect - Hair (Figure 4c)
```{r}
se_krtap <- 
  se[c("KRTAP5-8",
       "KRTAP5-3",
       "KRTAP10-4",
       "KRTAP5-11",
       "KRTAP5-7",
       "KRTAP5-10",
       "KRT71",
       "KRT27"),]

KRTAP_filt_d <- 
  se_krtap %>% assay(2) %>% log1p() %>% t() %>% scale() %>% t()

sample_site_color <- 
  colData(se_krtap) %>% as_tibble() %>% pull(biopsy_area) %>% unique()
color <- RColorBrewer::brewer.pal(length(sample_site_color), "Set3")
names(color) <- sample_site_color

heatmap_krtap_annotation <- 
  HeatmapAnnotation(
    `Tissue type` = colData(se_krtap) %>% as_tibble() %>% pull(skin_type), 
    `Anatomic region` = colData(se_krtap) %>% as_tibble() %>% pull(biopsy_area),
    col = list(`Tissue type` = pca_color,
               `Anatomic region` = color),
    # simple_anno_size = unit(1, "cm"),
    annotation_legend_param = list(`Tissue type` = list(nrow = 1, 
                                                        # grid_height = unit(2, "cm"), 
                                                        # grid_width = unit(1, "cm"), 
                                                        title_gp = gpar(fontsize = 16),
                                                        labels_gp = gpar(fontsize = 16)),
                                   `Anatomic region` = list(nrow = 2,
                                                          # grid_height = unit(2, "cm"), 
                                                          # grid_width = unit(1, "cm"), 
                                                          title_gp = gpar(fontsize = 16),
                                                          labels_gp = gpar(fontsize = 16))
                                   ))
# grid_height = unit(1, "cm"), grid_width = unit(5, "mm")
                                               
heatmap_krtap <- ComplexHeatmap::Heatmap(KRTAP_filt_d,
                                         name = "gene expression",
                                         column_dend_reorder = TRUE,
                        # clustering_method_columns = "complete",
                        show_column_names = FALSE,
                        show_row_dend = FALSE,
                        row_names_gp = gpar(fontsize = min(12, 800 * 1.5 / dim(KRTAP_filt_d)[1])),
                        top_annotation = heatmap_krtap_annotation,
                        heatmap_legend_param = list(
                          # title_gp = gpar(fontsize = 14),
                          # labels_gp = gpar(fontsize = 12)
                          ))

draw(heatmap_krtap, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)
```

```{r output figure 4c, eval=FALSE}
png("data/supplementary/figure_4c.png", 
    width = 480 * 5 , 
    height = 480 * 5 / 1.5, 
    res = 300)
draw(heatmap_krtap, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)
dev.off()
```

## Cell deconvolution (Figure S4, Table 2)
```{r MuSic, cache=TRUE}
# Construct ExpressionSet for bulk RNA-seq
se_scaled <- se
assay(se_scaled) <- NULL
multiomics <- as(se_scaled, "ExpressionSet")

# GSE121212
se_gse121212 <-
  read_rds("../gse121212/data/rds/se.rds") # download from https://github.com/tuhulab/gse121212/blob/master/data/rds/se.rds

assay(se_gse121212) <- NULL
colnames(se_gse121212) <- 
  colnames(se_gse121212) %>%
  str_remove_all("/home/projects/ku_00015/data/gse121212_data/mapping/hisat2/|\\.sra|\\.bam")
gse121212 <- as(se_gse121212, "ExpressionSet")

# Construct ExpressionSet for scRNA-seq
emma_guttman <- readRDS("data/ext/emma_scrna_ExpressionSet.rds")
multiomics_music <- music_prop(bulk.eset = multiomics, 
                               sc.eset = emma_guttman,
                               clusters = "cell_type", 
                               samples = "sample_name")
gse121212_music <-  music_prop(bulk.eset = gse121212, 
                               sc.eset = emma_guttman,
                               clusters = "cell_type", 
                               samples = "sample_name")
multiomics_results <-
  data.matrix(multiomics_music$Est.prop.weighted) %>% 
  as.data.frame() %>% rownames_to_column("BAM_ID") %>% 
  pivot_longer(!BAM_ID, names_to = "cell_type", values_to = "prop") %>% 
  left_join(colData(se_scaled) %>% as_tibble() %>% select(BAM_ID, group, subject, visit, skin_type, replicate_ID, gender, scorad, scorad_objective, easi_total_score, date_visit, rna_quality, visit_quarter, biopsy_area))
# saveRDS(multiomics_results, "data/MuSiC_result.rds")

gse121212_results <- 
  data.matrix(gse121212_music$Est.prop.weighted) %>% 
  as.data.frame() %>% rownames_to_column("SRR_ID") %>% 
  pivot_longer(!SRR_ID, names_to = "cell_type", values_to = "prop") %>% 
  left_join(
    colData(se_gse121212) %>% as_tibble %>% 
      select(SRR_ID, group, subject_id, skin_type, Trait, Sex, Age) %>% 
      mutate(skin_type = 
               case_when(skin_type == "lesional" ~ "LS",
                         skin_type == "non-lesional" ~ "NL",
                         skin_type == "healthy" ~ "HC",
                         skin_type == "chronic_lesion" ~ "LS"))) %>% 
  filter(Trait != "PSO")

# Result table

gse121212_multiomics_cell_type_table <- 
  gse121212_results %>% 
  group_by(skin_type, cell_type) %>% 
  summarise(mean = mean(prop),
            sd = sd(prop)) %>% 
  mutate(study = "gse121212") %>% 
  bind_rows(
    multiomics_results %>% 
      group_by(skin_type, cell_type) %>% 
      summarise(mean = mean(prop),
                sd = sd(prop)) %>% 
      mutate(study = "GENAD")) %>% 
  mutate(
    mean = scales::label_percent(accuracy = .1)(mean),
    sd = scales::label_percent(accuracy = .1)(sd),
    value = paste0(mean, "Â±", sd)) %>% 
  select(-mean, -sd) %>% 
  pivot_wider(names_from = cell_type,
              values_from = value)

gse121212_multiomics_cell_type_table %>% 
  write_tsv("data/gse121212_genad_cell_type.tsv")

cell_type_rank <- 
  multiomics_results %>% group_by(cell_type) %>% 
  summarize(mean_prop = mean(prop)) %>% arrange(-mean_prop) %>% pull(cell_type)

exclude_low_KRT <-
  multiomics_results %>% 
  filter(cell_type == "Keratinocytes", prop < .11) %>% pull(BAM_ID)

multiomics_results <-
  multiomics_results %>% filter(!BAM_ID %in% exclude_low_KRT) %>% 
  mutate(cell_type = factor(cell_type, levels = cell_type_rank),
         prop = prop * 100)

multiomics_results %>% 
  group_by(skin_type, cell_type) %>% 
  summarise(mean = mean(prop)) %>% 
  ggbarplot("skin_type", "mean", fill = "cell_type",
            xlab = "Tissue type", ylab = "Proportion(%)", palette = "npg") %>% 
    ggpar(legend = "bottom", 
        legend.title = "Cell type",
        font.legend = 7, 
        font.tickslab = 6)
```

```{r output figure s4, eval=FALSE}
multiomics_results_sample <- 
  multiomics_results %>%
  mutate(plot_name = paste(subject, visit, skin_type, replicate_ID, sep = "_"))

multiomics_results_sample_count <-
  multiomics_results_sample$plot_name %>% table

multiomics_results_sample <-
  multiomics_results_sample %>% 
  filter(plot_name %in% names(which(multiomics_results_sample_count == 10)))

# multiomics_results_sample %>% 
#   saveRDS("data/_TMP/cell_type_deconvolution.rds")

multiomics_results_vis_sample <- 
  multiomics_results_sample %>%  # filter(subject %in% c("AD_16", "CO_33")) %>% 
  arrange(plot_name) %>% 
  ggbarplot(x = "plot_name", y = "prop", fill = "cell_type", 
            xlab = "Sample", ylab = "Proportion(%)", palette = "npg") %>% 
  facet(facet.by = "subject", scales = "free") %>% 
  ggpar(xtickslab.rt = 90, font.tickslab = 6)

pdf("data/supplementary/figure_s4.pdf", height = 16.5 * 1.5, width = 11.7)
multiomics_results_vis_sample
dev.off()
```

### Statistical testing mean change (Table 2)

#### ALT
```{r ALT}
MuSiC_matrix <-
  data.matrix(multiomics_music$Est.prop.weighted)
ALT <- # addictive log-ratio transformation
  MuSiC_matrix %>% 
  as_tibble(rownames = "BAM_ID") %>% 
  # filter(Keratinocytes > .11) %>% 
  mutate(across(!BAM_ID, ~ ifelse(.x == 0, NA, .x)),
         across(!c("BAM_ID", "Melanocytes"), 
                ~ log2(.x / Melanocytes)))  %>% 
  left_join(
    colData(se_scaled) %>% as_tibble() %>% 
      select(BAM_ID, group, subject, visit, 
             skin_type, replicate_ID, gender, 
             scorad, scorad_objective, easi_total_score, 
             date_visit, rna_quality, visit_quarter, biopsy_area)) %>% 
  rename(T_cells = `T-cells`)
lmm_test <- 
  function(index, data, cell_type_n){
    f <- paste(cell_type_n[index], "~", "skin_type", "+",
               "(1|subject)", "+", "visit_quarter", "+",
               "biopsy_area")
    lmm_nonpaired <-
      lmer(f %>% as.formula(), data) %>% tidy() %>%
      filter(str_detect(term, "skin_type")) %>%
      mutate(term = paste(str_remove(term, "skin_type"), "vs HC"))
         
    lmm_paired <-
      lmer(f %>% as.formula(), 
           data %>%
             filter(skin_type %in% c("LS", "NL"))) %>%
      tidy()  %>%
      filter(str_detect(term, "skin_type")) %>%
      mutate(term = paste(str_remove(term, "skin_type"), "vs NL"))

    lmm_t <- bind_rows(lmm_nonpaired,
                       lmm_paired) %>%
      mutate(cell_type = cell_type_n[index])
  }


ALT_m <- 
  lapply(seq_along(c(colnames(ALT)[2:4], colnames(ALT)[6:11])), 
         lmm_test, data = ALT, 
         cell_type_n = c(colnames(ALT)[2:4], colnames(ALT)[6:11]))

ALT_p <- 
  ALT_m %>% map_dfr(~ .x) %>% 
  select(term, p.value, cell_type) %>% 
  mutate(p.value = p.value %>% round(2)) %>% 
  pivot_wider(names_from = term, values_from = p.value)

ALT_estimate <-
  ALT_m %>% map_dfr(~ .x) %>% 
  select(term, estimate, cell_type) %>% 
  mutate(estimate = estimate %>% round(2)) %>% 
  pivot_wider(names_from = term, values_from = estimate)
```

#### CLT
```{r statisticcs}
C_v <- MuSiC_matrix %>% colMeans() # column mean vector
C_m <- # column mean matrix
  rep(C_v, nrow(MuSiC_matrix)) %>% matrix(nrow = nrow(MuSiC_matrix), byrow = T) 
N_m <- # normalized matrix
  log2(data.matrix(multiomics_music$Est.prop.weighted) / C_m)

CLT <- # centred log-ratio transformation
  N_m %>% 
  as_tibble(rownames = "BAM_ID") %>% 
  mutate(across(!BAM_ID, ~ ifelse(.x == -Inf, NA, .x)))  %>% 
  left_join(
    colData(se_scaled) %>% as_tibble() %>% 
      select(BAM_ID, group, subject, visit, 
             skin_type, replicate_ID, gender, 
             scorad, scorad_objective, easi_total_score, 
             date_visit, rna_quality, visit_quarter, biopsy_area)) %>% 
  rename(T_cells = `T-cells`)

# CLT

CLT_m <- 
  lapply(seq_along(colnames(CLT)[2:11]), lmm_test, data = CLT, 
         cell_type_n = colnames(CLT)[2:11])
CLT_p <- 
  CLT_m %>% map_dfr(~ .x) %>% 
  select(term, p.value, cell_type) %>% 
  mutate(p.value = p.value %>% round(2)) %>% 
  pivot_wider(names_from = term, values_from = p.value)
CLT_estimate <-
  CLT_m %>% map_dfr(~ .x) %>% 
  select(term, estimate, cell_type) %>% 
  mutate(estimate = estimate %>% round(2)) %>% 
  pivot_wider(names_from = term, values_from = estimate)
```

### Diversity measure
```{r diversity measure}
shannon <- 
  (data.matrix(multiomics_music$Est.prop.weighted) * 100) %>% 
  vegan::diversity()

diversity <-
  colData(se_scaled) %>% as_tibble() %>% 
  select(BAM_ID, group, subject, visit, 
         skin_type, replicate_ID, gender, 
         scorad, scorad_objective, easi_total_score, 
         date_visit, rna_quality, 
         visit_quarter, biopsy_area) %>% 
  mutate(shannon = shannon)

diversity_LSvsNL_lmm <- 
  lmer(shannon ~ 
       skin_type + (1|subject) + biopsy_area, 
       diversity %>% filter(skin_type %in% c("LS", "NL"))) %>% 
  tidy() %>% 
  filter(term %>% str_detect("skin_type"))

diversity_LS_NLvsHC_lmm <- 
  lmer(shannon ~ 
       skin_type + (1|subject) + 
         (1|visit_quarter) + biopsy_area, 
       diversity) %>% 
  tidy() %>% 
  filter(term %>% str_detect("skin_type"))

```

```{r dominance and evenness}
dominance <- 
  (data.matrix(multiomics_music$Est.prop.weighted) * 100) %>% 
  as_count() %>% 
  index_heterogeneity(method = "simpson")

evenness <-
  (data.matrix(multiomics_music$Est.prop.weighted) * 100) %>% 
  as_count() %>% 
  index_evenness()

similarity <-
  (data.matrix(multiomics_music$Est.prop.weighted) * 100) %>% 
  as_count() %>% 
  similarity(method = "bray")

diversity_t <- 
  colData(se) %>% as_tibble() %>% 
  mutate(dominance = dominance@values,
         evenness = evenness@values)

lmer(dominance ~ skin_type + 
       biopsy_area +
       (1|visit) + (1|subject), 
     diversity_t) %>% 
  tidy()

lmer(evenness ~ skin_type + 
       biopsy_area +
       (1|visit) + (1|subject), 
     diversity_t) %>% 
  tidy()

diversity_t %>% ggplot(aes(skin_type, dominance)) + geom_point() + geom_boxplot()
diversity_t %>% ggplot(aes(skin_type, evenness)) + geom_point() + geom_boxplot()
diversity_t %>% ggplot(aes(dominance, evenness)) + geom_point() + geom_smooth(method = "lm")
```

### Correlation between Schwaan cells and SCORAD
```{r cor}
ALT_scorad_t <-
  ALT %>%
  left_join(extensive_meta %>%
              select(BAM_ID, contains("scorad"), -scorad, -scorad_objective) %>%
              mutate(across(!BAM_ID, ~ as.numeric(.x))),
            by = "BAM_ID") %>%
  filter(skin_type == c("LS"))  %>%
  column_to_rownames("BAM_ID")
cor.test(x = ALT_scorad_t$scorad, 
         y = ALT_scorad_t$Neuron_and_Schwann_cells, 
         use = "complete.obs", method = "spearman") %>% tidy()
cor.test(x = ALT_scorad_t$scorad_pruritus_vas_0_10, 
         y = ALT_scorad_t$Neuron_and_Schwann_cells, 
         use = "complete.obs", method = "spearman") %>% tidy()
cor.test(x = ALT_scorad_t$scorad_sleep_loss_vas_0_10, 
         y = ALT_scorad_t$Neuron_and_Schwann_cells, 
         use = "complete.obs", method = "spearman") %>% tidy()
cor.test(x = ALT_scorad_t$scorad_lichenification, 
         y = ALT_scorad_t$Neuron_and_Schwann_cells, 
         use = "complete.obs", method = "spearman") %>% tidy()
```

## Genome regulatory elements (Figure S2)

### Transcription factor enrichment (Figure S2a)
```{r gsea - regulatory elements}

# gene set regulatory
gsea_r_p <- 
  gsea_res_filt %>% filter(pathway %in% gs_r, 
                           !contrast %>% str_detect("PSO"),
                           padj < .05,
                           !contrast %in% c("Adult_LSvsAdult_NN",
                                            "Adult_NLvsAdult_NN",
                                            "Adult_NLvsAdult_NN",
                                            "Children_NLvsChildren_NN",
                                            "Children_LSvsChildren_NN",
                                            "Adult_LSvsChildren_LS",
                                            "Adult_NLvsChildren_NL",
                                            "LS/NN_derm",
                                            "LS/NN_epi",
                                            "LS/NN_FT",
                                            "NL/NN_derm",
                                            "NL/NN_FT")) %>%
  mutate(project = ifelse(project == "multi-omics",
                          "GENAD",
                          project),
         project = factor(project, levels = c("gse121212",
                                         "acute_chronic",
                                         "gse107361",
                                         "gse120721",
                                         "GENAD"))) %>% 
  ggplot(aes(contrast, des, size = `-log10p`, color = NES)) + 
  geom_point() +
  scale_size_area(max_size = 6) + 
  scale_color_gradient2(low="blue", mid = "white", high="red") +
  theme_bw() +
  facet_grid(~project, scales = "free", space = "fixed") +
  theme(axis.title = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1))
ggsave("data/supplementary/figure_s2_a.png", gsea_r_p, 
       width = 6.5,
       height = 3.5,
       dpi = 300)
```

```{r figure 5c boxplot - ncRNA}
toptopncRNA <- 
  vp_core_t %>% 
  filter(gene_name %>% str_detect("LINC"),
         skin_type > .05) %>% 
  mutate(group = gene_name %>% str_extract("LINC")) %>%
  group_by(group) %>% 
  slice_max(skin_type, n = 5) %>% pull(gene_name)

toptopncRNA_boxplot <-
  se[toptopncRNA,] %>% as_tibble() %>% 
  mutate(feature = feature %>% forcats::fct_relevel(toptopncRNA)) %>% 
  ggboxplot(x = "skin_type", y = "counts_scaled", add = "jitter", 
            facet.by = "feature", xlab = FALSE,
            ylab = FALSE, color = "skin_type", 
            palette = pca_color, legend.title = "",
            font.legend = c(14), 
            nrow = 1, 
            yscale = "log10")

ggsave("data/supplementary/figure_s2_b.png", 
       toptopncRNA_boxplot, 
       width = 6.5,
       height = 3.5,
       dpi = 300)
```

#### LINC 
```{r}
LINC_g <- 
  de_v1_c %>% pull(gene_name) %>% 
  grep("LINC", ., value = T) %>% unique

MIR_g <- 
  de_v1_c %>% pull(gene_name) %>% 
  grep("MIR", ., value = T) %>% unique
```

```{r heatmap-linc}
se_linc <- se[LINC_g,]
heatmap_linc_data <- 
  se_linc 

heatmap_linc_data_m <-
  scale(assay(heatmap_linc_data, 2) %>% t() %>% log1p(), 
        center = TRUE, scale = TRUE) %>% t()
sample_site_color <- 
  colData(heatmap_linc_data) %>% as_tibble() %>% pull(biopsy_area) %>% unique()
color <- RColorBrewer::brewer.pal(length(sample_site_color), "Pastel2")
names(color) <- sample_site_color
heatmap_linc_annotation <- 
  HeatmapAnnotation(
    `Tissue type` = colData(heatmap_linc_data) %>% as_tibble() %>% pull(skin_type), 
    `Anatomic region` = colData(heatmap_linc_data) %>% as_tibble() %>% pull(biopsy_area),
    col = list(`Tissue type` = pca_color,
               `Anatomic region` = color),
    annotation_legend_param = list(`Tissue type` = list(nrow = 1, 
                                                        title_gp = gpar(fontsize = 16),
                                                        labels_gp = gpar(fontsize = 16)),
                                   `Anatomic region` = list(nrow = 1,
                                                          title_gp = gpar(fontsize = 16),
                                                          labels_gp = gpar(fontsize = 16))
                                   ))

library(dendsort)
col_dend <- heatmap_linc_data_m %>% t() %>% dist() %>% hclust
col_dend_sort <- dendsort(col_dend)

heatmap_linc <- ComplexHeatmap::Heatmap(heatmap_linc_data_m,
                        name = "gene expression",
                        column_dend_reorder = TRUE,
                        row_dend_reorder = TRUE,
                        show_column_names = FALSE,
                        show_row_names = T,
                        show_row_dend = FALSE,
                        # cluster_columns = col_dend_sort,
                        row_names_gp = gpar(fontsize = min(8, 800 * 6 /dim(heatmap_linc_data)[1])),
                        top_annotation = heatmap_linc_annotation,
                        heatmap_legend_param = list(
                          title_gp = gpar(fontsize = 16),
                          labels_gp = gpar(fontsize = 14)
                          ))

draw(heatmap_linc, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)
```

```{r}
png("data/supplementary/figure_s5.png", 
    width = 480*6, height = 480*11, res=300)
draw(heatmap_linc, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)
dev.off()
```

```{r heatmap-mir}
se_mir <- se[MIR_g,]
heatmap_mir_data <- 
  se_mir
heatmap_mir_data_m <-
  scale(assay(heatmap_mir_data, 2) %>% t() %>% log1p(), 
        center = TRUE, scale = TRUE) %>% t()
sample_site_color <- 
  colData(heatmap_mir_data) %>% as_tibble() %>% pull(biopsy_area) %>% unique()
color <- RColorBrewer::brewer.pal(length(sample_site_color), "Pastel2")
names(color) <- sample_site_color
heatmap_mir_annotation <- 
  HeatmapAnnotation(
    `Tissue type` = colData(heatmap_mir_data) %>% as_tibble() %>% pull(skin_type), 
    `Anatomic region` = colData(heatmap_mir_data) %>% as_tibble() %>% pull(biopsy_area),
    col = list(`Tissue type` = pca_color,
               `Anatomic region` = color),
    # simple_anno_size = unit(1, "cm"),
    annotation_legend_param = list(`Tissue type` = list(nrow = 1, 
                                                        # grid_height = unit(2, "cm"), 
                                                        # grid_width = unit(1, "cm"), 
                                                        title_gp = gpar(fontsize = 16),
                                                        labels_gp = gpar(fontsize = 16)),
                                   `Anatomic region` = list(nrow = 1,
                                                          # grid_height = unit(2, "cm"), 
                                                          # grid_width = unit(1, "cm"), 
                                                          title_gp = gpar(fontsize = 16),
                                                          labels_gp = gpar(fontsize = 16))
                                   ))


heatmap_mir <- ComplexHeatmap::Heatmap(heatmap_mir_data_m,
                        name = "gene expression",
                        # column_dend_reorder = TRUE,
                        # row_dend_reorder = TRUE,
                        # clustering_method_columns = "complete",
                        cluster_columns = heatmap_mir_data_m %>% t() %>% dist() %>% hclust() %>% dendsort(),
                        cluster_rows = heatmap_mir_data_m %>% dist() %>% hclust() %>% dendsort(),
                        show_column_names = FALSE,
                        show_row_names = T,
                        show_row_dend = FALSE,
                        row_names_gp = gpar(fontsize = 10),
                        top_annotation = heatmap_mir_annotation,
                        heatmap_legend_param = list(
                          title_gp = gpar(fontsize = 16),
                          labels_gp = gpar(fontsize = 14)
                          ))

draw(heatmap_mir, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)
```

```{r save heatmap}
png("data/supplementary/figure_s2_c.png", 
    width = 480*6.5, 
    height = 480*4, 
    res=300)
draw(heatmap_mir, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)
dev.off()
```

```{r MIR31 and CLDN1}
MIR31_CLDN1 <- 
  se[c("MIR31HG", "CLDN1"),] %>% as_tibble() %>% 
  filter(skin_type == "LS") %>% 
  select(feature, sample, counts_scaled, skin_type) %>% pivot_wider(names_from = feature, values_from = counts_scaled) %>% 
  ggscatter(x = "CLDN1", 
            y = "MIR31HG",
            ylab = "MIR31HG",
            xscale = "log10",
            add = "reg.line",
            ) +
  stat_cor()
ggsave("data/supplementary/figure_s2_d.png", 
       MIR31_CLDN1, width = 7, height = 5, dpi = 300)
```


## Disease severity change
### Figure S6
```{r}
extensive_meta <- readr::read_csv("../multiomics-ad-transcriptomics//data/RNAseq_sample_annotation(extensive).csv")
time_course_ad_subject <- 
  extensive_meta %>% filter(visit == "02", group == "AD") %>% pull(subject) %>% unique

disease_severity <- extensive_meta %>% 
  dplyr::filter(subject %in% time_course_ad_subject) %>% 
  dplyr::select(-skin_type) %>% 
  distinct() %>% 
  dplyr::select(visit, visit_quarter, subject, 
                easi_total_score, scorad_objective, scorad) %>% 
  tidyr::pivot_longer(cols = c("easi_total_score",
                               "scorad_objective",
                               "scorad"), 
                      names_to = "disease_severity") %>% 
  filter(subject != "AD_11") %>% 
  mutate(disease_severity = 
            case_when(disease_severity == "easi_total_score" ~ "EASI",
                      disease_severity == "scorad" ~ "SCORAD",
                      disease_severity == "scorad_objective" ~ "oSCORAD"),
         disease_severity = forcats::fct_relevel(disease_severity,
                                                 c("EASI", 
                                                   "SCORAD",
                                                   "oSCORAD")),
         visit = visit %>% as.character(),
         visit_quarter = visit_quarter %>% as.character())


```

```{r}
disease_severity_natural_year <- extensive_meta %>% 
  dplyr::filter(subject %in% time_course_ad_subject) %>% 
  dplyr::select(-skin_type) %>% 
  distinct() %>% 
  mutate(year = date_visit %>% format("%Y"),
         year_quarter = paste0(year, visit_quarter)) %>% 
  dplyr::select(visit, visit_quarter, year_quarter, subject, 
                easi_total_score, scorad_objective, scorad) %>% 
  tidyr::pivot_longer(cols = c("easi_total_score",
                               "scorad_objective",
                               "scorad"), 
                      names_to = "disease_severity") %>% 
  filter(subject != "AD_11") %>% 
  mutate(disease_severity = 
            case_when(disease_severity == "easi_total_score" ~ "EASI",
                      disease_severity == "scorad" ~ "SCORAD",
                      disease_severity == "scorad_objective" ~ "oSCORAD"),
         disease_severity = forcats::fct_relevel(disease_severity,
                                                 c("EASI", 
                                                   "SCORAD",
                                                   "oSCORAD")),
         visit = visit %>% as.character(),
         visit_quarter = visit_quarter %>% as.character())

disease_severity_natural_year_g <- 
  disease_severity_natural_year %>% 
  ggplot(aes(year_quarter, value, group = subject, color = subject)) +
  geom_point() + geom_line() + facet_grid(disease_severity~., scales = "free_y") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  xlab("Time") +
  theme_classic() +
  theme(axis.title.y =  element_blank())

ggsave("data/supplementary/figure_s7.png", disease_severity_natural_year_g)
```



```{r baseline characteristics}
baseline_disease_severity <- 
  clinical_records %>% filter(visit_no == 1, group == "AD") %>% 
  summarise(mean_scorad = mean(scorad, na.rm = TRUE) %>% round(1),
            mean_easi = mean(easi_total_score) %>% round(2))
```

```{r follow-up subjects}
follow_up <- 
  clinical_records %>% filter(visit_no != 1) %>% 
  group_by(group) %>% 
  select(id, group) %>% 
  distinct() %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = group, values_from = n)
```





```{r test}
tibble(
  visit = c(rep("visit 1", 10), rep("visit 2", 10), rep("visit 3", 10)),
  count = rnorm(n = 30)
) %>% 
  group_by(visit) %>% 
  mutate(group_mean = mean(count),
         outlier = ifelse( (count - group_mean) > (2* abs(group_mean)), 1, .5)) %>% 
  ggplot(aes(visit, count)) +
  geom_dotplot(aes(alpha = outlier), binaxis = "y", stackdir = "center")
```

