---
title: "03-MOMIX GENAD"
author: "Tu Hu"
date: "4/5/2022"
output: ''
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(tidySummarizedExperiment)
library(ggpubr)
```

## Load data

### AD risk SNPs

```{r Phenome}

```


```{r Genetic Risk}
genetic_risk <-
  readr::read_rds("../multiomics-ad-genome-a4086/data/genetic_risk_dosage.rds") %>% 
  tibble::column_to_rownames("SNP")

# query SNP info
SNP_info <-
  genetic_risk %>% rownames() %>% 
  gprofiler2::gconvert()

# shape into individual
genetic_risk_SNP <-
  genetic_risk %>% tibble::rownames_to_column("SNPid") %>% 
  pivot_longer(cols = !SNPid, names_to = "subject",
               values_to = "genetic_risk_dosage") %>% 
  mutate(group = subject %>% stringr::str_extract("AD|CO"))
```

```{r Gene expression}
gene_expression <-
  readr::read_rds("data/se.rds")
```

```{r Proteome}

```

```{r Metabolome}
metabolites <- 
  readr::read_rds("../multiomics-ad-metabolome-life/data/metabolome_data_wide.rds.gz")
```

```{r Microbiome}

```


## Logistic regression
```{r}
gene_expression <-
  readr::read_rds("data/se.rds")

IL22_expr <- 
  gene_expression["IL22", ] %>% 
  as_tibble()

IL22_expr %>% 
  ggboxplot(x = "skin_type",
            y = "counts_scaled",
            add = "jitter",
            xlab = "Tissue type",
            color = "skin_type",
            palette = c("LS" = "#eb2d0c", "NL" = "#eb8b9b", "HC" = "#91cf60"),
            title = "IL22 expr")

# whether genetic risk add IL22 expr

IL22_expr_t_fit <-
  IL22_expr %>% 
  filter(skin_type != "NL") %>% 
  rename(IL22_expr = counts_scaled) %>% 
  mutate(IL22_expr = log2(IL22_expr + .5),
         group = ifelse(group == "CO", 0, 1)) %>% 
  left_join(genetic_risk_SNP %>% 
              filter(SNPid == "rs10791824") %>% 
              select(-group), by = "subject")
  
glm(
  group ~ IL22_expr + genetic_risk_dosage, 
  family = "binomial",
  IL22_expr_t_fit
) %>% summary()
```

## MOFA





