# Skin transcriptome data cleaning

## Data cleaning
```{r initiate data cleaning, echo=FALSE, message=FALSE, warning=FALSE}
pacman::p_load(tibble, dplyr, tidyr, readr, stringr, ggplot2, purrr, 
               tidybulk, tidySummarizedExperiment, 
               ComplexHeatmap, ggpubr,
               BiocParallel, Biobase, variancePartition,
               DESeq2, limma, edgeR, MuSiC, convert, 
               patchwork)
knitr::opts_chunk$set(message = FALSE, warning = FALSE, echo = FALSE, fig.align = 'center')
pca_color <- c("LS" = "#eb2d0c", "NL" = "#eb740c", "NN" = "#91cf60")
core_n <- future::availableCores()
register(MulticoreParam(ifelse(core_n <= 8, core_n - 2, core_n -6)))
```

```{r load data}
# read raw counts ---------------------------------------
counts <-
  data.table::fread("../multiomics-ad-transcriptomics/data/counts/counts_gene_name_redo.txt.gz") %>% 
  mutate(Chr = Chr %>% str_extract("chr\\w{1,}(?=;)|chr\\w{1,}$"), #keeps only 1
         Strand = Strand %>% str_extract("\\+|\\-"),               #keeps only 1
         Start = Start %>% str_extract("\\d{1,}(?=;)|\\d{1,}$"),   #keeps only 1
         End = End %>% str_extract("\\d{1,}(?=;)|\\d{1,}$")) %>%   #keeps only 1
  filter(!Geneid %>% 
           stringr::str_detect("\\d{1,}P$|\\d{1,}P\\d{1,}$|\\.|-AS\\d{1}|-DT")) %>% # apply gene filtering (pseudo genes, antisense genese, DT genes)
  dplyr::rename_with(str_extract, starts_with("/home/projects/"), pattern = "NG-[:graph:]{1,}$")

# lib to merge -------------
lib_id <- colnames(counts)[-1:-6] %>% str_extract("lib\\d{1,}") # extract lib id
lib_id_t <- table(lib_id)
lib_id_merge <- lib_id_t[(lib_id_t != 1)] %>% rownames() #lib_ids to be merged

# merge technical replicates -------------------------------------------
source("../multiomics-ad-transcriptomics/R/helper.R")

counttable_LibMerged <- 
  bind_cols(counts[, 1:6], 
            counttable_merge_library_fun(counts %>% dplyr::select(-1:-6), lib_id_merge)) 

# clinical records ------------------------
clinical_records <- 
  readr::read_csv("../multiomics-ad-transcriptomics/data/LEO-AD-metadata.csv") %>% 
  mutate(id = id %>% str_replace("-", "_"),
         JOIN_ID = paste(id, paste0("0",visit_no), sep="_"))

extensive_meta <-
  readr::read_csv("../multiomics-ad-transcriptomics/data/RNAseq_sample_annotation(extensive).csv")
```

```{r SummarizedExperiment}
# revised biopsy area by Tanja
revised_biopsy_area <- 
  readr::read_csv("../multiomics-ad-transcriptomics/data/tanja_replicate_location.csv") %>% 
  mutate(biopsy_area = biopsy_area %>% str_replace_all(" ", "_"))

# revise space infomation
biopsy_area <- 
  extensive_meta %>% 
  select(BAM_ID, biopsy_area_old = biopsy_area) %>% 
  left_join(revised_biopsy_area %>% select(BAM_ID, biopsy_area)) %>% 
  mutate(biopsy_area = ifelse(is.na(biopsy_area), biopsy_area_old, biopsy_area)) %>% 
  mutate(biopsy_area = ifelse(biopsy_area == "lÃ¦g", "leg", biopsy_area)) %>% 
  select(-biopsy_area_old) %>% 
  add_case(BAM_ID = "NG-23830_01_AD_07_BI_LS_01_E9_lib390881_6751_1.bam",
           biopsy_area = "arm")

# se col_data
col_data <-
  tibble(
    # definingn metadata --------------------------------------
    BAM_ID = colnames(counttable_LibMerged)[-1:-6]) %>% 
    mutate(sequencing_batch = BAM_ID %>% str_extract("NG-\\d{1,}"),
           group = BAM_ID %>% str_extract("AD|CO") %>% 
               forcats::fct_relevel(c("CO", "AD")),
           subject = BAM_ID %>% str_extract("(AD|CO)_\\d{2,}"),
           visit = BAM_ID %>% str_extract("\\d{2,}_(AD|CO)") %>% str_extract("\\d{2,}"),
           skin_type = BAM_ID %>% str_replace("TS", "LS") %>% str_extract("NN|NL|LS") %>% forcats::fct_relevel(c("NN", "NL", "LS")),
           replicate_ID = BAM_ID %>% str_replace("TS_BI", "BI_LS") %>% str_extract("(NN|NL|LS)_\\d{2,}") %>% str_extract("\\d{2,}"),
           library_ID = BAM_ID %>% str_extract("lib\\d{1,}"),
           sequencing_id = BAM_ID %>% str_extract("\\d{4}_(1|2)"),
           JOIN_ID = base::paste(subject, visit, sep = "_")) %>%
    left_join(clinical_records %>% select(id, gender) %>% filter(!is.na(gender)), by=c("subject"="id")) %>%
    left_join(clinical_records %>% select(JOIN_ID, scorad, scorad_objective, easi_total_score, date_visit)) %>% 
    select(-JOIN_ID) %>% 
    mutate(gender = gender %>% forcats::fct_relevel(c("female", "male"))) %>% 
    left_join(extensive_meta %>% select(BAM_ID, rna_quality, visit_quarter)) %>% 
    left_join(biopsy_area)

# Summarized Experiment--------------------
se <- SummarizedExperiment::SummarizedExperiment(
  assays = list(counts = counttable_LibMerged %>% select(-1:-6) %>% as.matrix),
  colData = col_data
)
names(se) <- counttable_LibMerged$Geneid
```

```{r filtering}
# filter out samples ----------------
se <- se[, se$BAM_ID != (assay(se) %>% colSums() %>% which.min %>% names)] # 392
se <- se[, !se$library_ID %in% c("lib390174", "lib390716")] # 390
```

```{r revise some sp info}
index_ad03v01ls01 <-
  which(colData(se)$subject == "AD_03" &
        colData(se)$visit == "01" &
        colData(se)$skin_type == "LS" &
        colData(se)$replicate_ID == "01")
colData(se)$skin_type[index_ad03v01ls01] <- "NL"

index_ad03v01nl01 <-
  which(colData(se)$subject == "AD_03" &
          colData(se)$visit == "01" &
          colData(se)$skin_type == "NL" &
          colData(se)$replicate_ID == "01")
colData(se)$skin_type[index_ad03v01nl01] <- "LS"

index_ad13v03nl01 <-
  which(colData(se)$subject == "AD_13" &
        colData(se)$visit == "03" &
        colData(se)$skin_type == "NL" &
        colData(se)$replicate_ID == "01")
colData(se)$visit[index_ad13v03nl01] <- "05"
colData(se)$skin_type[index_ad13v03nl01] <- "LS"
colData(se)$replicate_ID[index_ad13v03nl01] <- "02"

index_ad13v05ls02 <-
  which(colData(se)$subject == "AD_13" &
        colData(se)$visit == "05" &
        colData(se)$skin_type == "LS" &
        colData(se)$replicate_ID == "02")
colData(se)$visit[index_ad13v05ls02] <- "03"
colData(se)$skin_type[index_ad13v05ls02] <- "NL"
colData(se)$replicate_ID[index_ad13v05ls02] <- "01"
```

```{r Sample QC}
sample_qc_record_edited <-
  readr::read_csv("data/sample_qc_record - sample_qc_record.csv")
```

```{r exclude}
exclude_BAM <- 
  colData(se) %>% as_tibble() %>% 
  select(BAM_ID, subject, visit, skin_type, replicate_ID) %>%
  left_join(sample_qc_record_edited) %>% 
  filter(exclude == TRUE) %>% pull(BAM_ID)
```

```{r generate new se}
se <- se[, !se$BAM_ID %in% exclude_BAM]
colData(se)$biopsy_area <- colData(se)$biopsy_area %>% 
  str_extract("arm|back_of_knee|elbow|feet|hand|leg|wrist")

se_prifilt <- se

# filter out genes 
se <- se %>% 
  keep_abundant(factor_of_interest = skin_type) %>% 
  scale_abundance()

```

```{r generate KRTAP}
se_KRTAP <- se_prifilt %>% 
  keep_abundant(factor_of_interest = skin_type, minimum_proportion = .1) %>%
  scale_abundance()
```

```{r KRTAP variance}
KRTAP_hair <-
  se_KRTAP %>% 
  filter(feature %in% c(paste0("KRTAP5-", 1:11), "KRTAP10-4")) %>% 
  group_by(feature) %>% 
  summarise(mean_expr = mean(counts), 
            sd_expr = sd(counts)) %>%   
  mutate(expr = ifelse(mean_expr >= 10, TRUE, FALSE))

se_KRTAP_filt <- 
  se_KRTAP %>% 
  filter(feature %in% (KRTAP_hair %>% filter(expr == TRUE) %>% pull(feature)))
```

```{r write out data, eval=FALSE}
readr::write_rds(se, "data/se.rds")
readr::write_rds(se_KRTAP_filt, "data/se_krtap.rds")
```

## Differential expressed genes: baseline

```{r DGE-V1, cache=TRUE}
# V1 LSvsNL --------------------
se_v1_LSvsNL <- se[, se$visit == "01" & se$skin_type %in% c("LS", "NL")]
de_v1_LSvsNL <- DESeq2::DESeqDataSet(se_v1_LSvsNL, ~ subject + skin_type)
de_v1_LSvsNL <- DESeq2::DESeq(de_v1_LSvsNL, parallel = TRUE)
de_v1_LSvsNL_LFC <- lfcShrink(de_v1_LSvsNL, coef = "skin_type_LS_vs_NL", type = "apeglm", parallel = TRUE, quiet = TRUE)
de_v1_LSvsNL_result <- de_v1_LSvsNL_LFC %>% as.data.frame() %>% tibble::rownames_to_column(var = "gene_name") %>% tibble()
de_v1_LSvsNL_result_s <- de_v1_LSvsNL_result %>% filter(padj < .05, abs(log2FoldChange) > 1)

# V1 LSvsNN --------------------
se_v1_LSvsNN <- se[, se$visit == "01" & se$skin_type %in% c("LS", "NN")]
de_v1_LSvsNN <- DESeq2::DESeqDataSet(se_v1_LSvsNN, ~ gender + skin_type) 
de_v1_LSvsNN <- DESeq2::DESeq(de_v1_LSvsNN, parallel = TRUE)
de_v1_LSvsNN_LFC <- lfcShrink(de_v1_LSvsNN, coef = "skin_type_LS_vs_NN", type = "apeglm", parallel = TRUE, quiet = TRUE)
de_v1_LSvsNN_result <- de_v1_LSvsNN_LFC %>% as.data.frame() %>% tibble::rownames_to_column(var = "gene_name") %>% tibble()
de_v1_LSvsNN_result_s <- de_v1_LSvsNN_result %>% filter(padj < .05, abs(log2FoldChange) > 1)

# V1 LSvsNN --------------------
se_v1_NLvsNN <- se[, se$visit == "01" & se$skin_type %in% c("NL", "NN")]
de_v1_NLvsNN <- DESeq2::DESeqDataSet(se_v1_NLvsNN, ~ gender + skin_type) 
de_v1_NLvsNN <- DESeq2::DESeq(de_v1_NLvsNN, parallel = TRUE)
de_v1_NLvsNN_LFC <- lfcShrink(de_v1_NLvsNN, coef = "skin_type_NL_vs_NN", type = "apeglm", parallel = TRUE, quiet = TRUE)
de_v1_NLvsNN_result <- de_v1_NLvsNN_LFC %>% as.data.frame() %>% tibble::rownames_to_column(var = "gene_name") %>% tibble()
de_v1_NLvsNN_result_s <- de_v1_NLvsNN_result %>% filter(padj < .05, abs(log2FoldChange) > 1)

# Consolidate the results -----
de_v1_c <-
  bind_rows(de_v1_LSvsNL_result_s %>% mutate(contrast = "LSvsNL"),
            de_v1_LSvsNN_result_s %>% mutate(contrast = "LSvsNN"),
            de_v1_NLvsNN_result_s %>% mutate(contrast = "NLvsNN")) %>% 
  mutate(direction = ifelse(log2FoldChange > 0, "up", "down"))
de_v1_c_s <-
  de_v1_c %>% 
  group_by(contrast, direction) %>% 
  summarise(n_dys = n())
# No of dysregulated genes
n_de_v1_LSvsNL <- de_v1_c_s %>% filter(contrast == "LSvsNL") %>% pull(n_dys)
n_de_v1_LSvsNN <- de_v1_c_s %>% filter(contrast == "LSvsNN") %>% pull(n_dys)
n_de_v1_NLvsNN <- de_v1_c_s %>% filter(contrast == "NLvsNN") %>% pull(n_dys)
de_v1_c %>% readr::write_csv("data/de_v1_sig.csv")
de_v1_full <- list(de_v1_LSvsNL = de_v1_LSvsNL_result, 
                   de_v1_LSvsNN = de_v1_LSvsNN_result, 
                   de_v1_NLvsNN = de_v1_NLvsNN_result)
# de_v1_full %>% readr::write_rds("data/de_v1_full.rds")
```

## Differential expressed genes: anotomic area
```{r DGE space, cache=TRUE}
DGE_space <-
  tibble(
    C1 = c("LS", "LS", "NL"),
    C2 = c("NL", "NN", "NN")
  ) %>% 
  mutate(
    contrast = paste0(C1, "vs", C2),
    subject_matched = ifelse(C1 == "LS" & C2 =="NL", TRUE, FALSE),
    se_exp = map2(C1, C2, ~ se[,se$skin_type %in% c(.x, .y)]),
    design_f = ifelse(subject_matched,
                      "~ subject + skin_type + visit + skin_type * visit + biopsy_area",
                      "~ gender + skin_type + visit + skin_type * visit + biopsy_area"),
    deseq = map2(se_exp, design_f, ~ DESeqDataSet(.x, .y %>% as.formula)),
    deseq = map(deseq, ~ DESeq(.x, parallel = T)),
    res_name = map(deseq, ~ resultsNames(.x) %>% grep(pattern = "biopsy_area", ., value = T))) %>% 
  unnest(cols = res_name) %>% 
  mutate(
    res_shrink = map2(deseq, res_name, ~ lfcShrink(.x, coef = .y, type = "apeglm", parallel = T)),
    res_shrink = map(res_shrink, ~ as_tibble(.x, rownames = "gene_name"))
  )
save(DGE_space %>% select(contrast, res_shrink), "data/dge_space.rds")
```

```{r DGE spcace, cache=TRUE}
se_s_LSvsNL <- 
  se_space %>% filter(skin_type %in% c("LS", "NL"))
de_s_LSvsNL <- 
  DESeq2::DESeqDataSet(se_s_LSvsNL, ~ subject + visit + skin_type + skin_type * visit + biopsy_area)
de_s_LSvsNL_m <- DESeq2::DESeq(de_s_LSvsNL, parallel = TRUE)

de_s_LSvsNL_legvsarm_res <- 
  bind_cols(tibble(gene_name = names(de_s_LSvsNL_m)),
  DESeq2::results(de_s_LSvsNL_m, name = "biopsy_area_leg_vs_arm") %>% as_tibble())

biopsy_area_contrast <-
  resultsNames(de_s_LSvsNL_m) %>% 
  str_extract("biopsy_area_[:graph:]{1,}") %>% 
  purrr::discard(is.na)

register(MulticoreParam(parallel::detectCores() - 1))

biopsy_area_results <- 
  lapply(biopsy_area_contrast, function(x){
  bind_cols(tibble(gene_name = names(de_s_LSvsNL_m)),
  lfcShrink(de_s_LSvsNL_m, 
            coef = x, 
            parallel = TRUE) %>% as_tibble() %>% 
    mutate(contrast = x))
}) %>% purrr::map_dfr(~.x) %>% filter(padj < .05, abs(log2FoldChange) > 1) %>% 
  mutate(contrast = contrast %>% str_remove("biopsy_area_"))
```

```{r DGE space - DT vis, eval=FALSE}
biopsy_area_results %>% 
  datatable() %>% 
  formatRound(2:4) %>% 
  formatSignif(5:6)
```

## Gene set enrichment analysis

```{r dge result}
de_v1_full <- readr::read_rds("data/de_v1_full.rds")
gse121212 <- readr::read_rds("https://rawcdn.githack.com/tuhulab/gse121212/0ba93240d13b753bf32fd0aab0841fb6968b0ea5/data/DGE_full_model.rds")
dge_result <-
  tibble(
    project = c(rep("multi-omics",length(de_v1_full)), 
                rep("gse121212",nrow(gse121212))),
    contrast = c(de_v1_full %>% names, gse121212$contrast),
    dge_result = c(de_v1_full, gse121212$res_shrink)
)
```


```{r prepare gene set, cache=TRUE}
# prepare gs --------------
gs <- 
  tibble(gs_name = c("CP:BIOCARTA", 
                     "CP:KEGG", 
                     "CP:PID", 
                     "CP:REACTOME", 
                     "CP:WIKIPATHWAYS", 
                     "MIR:MIRDB", 
                     "TFT:GTRD", 
                     "GO:BP", 
                     "GO:MF")) %>% 
  mutate(gs_list = map(gs_name, function(subcat){
  gs_df <- msigdbr::msigdbr(species = "Homo sapiens", 
                            subcategory = subcat) %>% 
    group_by(gs_name) %>% tidyr::nest() %>% 
    mutate(gene_id = purrr::map(data, ~ .x %>% pull(gene_symbol))) %>% select(-data)
  gs_list <- gs_df$gene_id
  names(gs_list) <- gs_df$gs_name
  return(gs_list)
}))
```

```{r fgsea, cache=TRUE}
gsea_exp <- 
  expand.grid(contrast = dge_result$contrast,
              gs_name = gs$gs_name) %>% 
  left_join(dge_result, by = "contrast") %>% 
  left_join(gs, by = "gs_name") %>% 
  mutate(rank = map2(dge_result, gs_name,
                     function(dge_result, gs_name){
                       p_cutoff <- ifelse(gs_name %in% c("MIR:MIRDB", 
                                                         "TFT:GTRD"),
                                          1, # .05
                                          1)
                       log2fc_cutoff <- ifelse(gs_name %in% c("MIR:MIRDB",
                                                        "TFT:GTRD"),
                                               0, # 1
                                               0)
                       dge_result %>% 
                         mutate(rank = log2FoldChange) %>% 
                         filter(padj < p_cutoff, abs(log2FoldChange) > log2fc_cutoff) %>% 
                         select(gene_name, rank) %>% deframe}),
         gse_res = map2(rank, gs_list,
                            function(rank, gs_list){
                              gsa_stat <- fgsea::fgsea(pathways = gs_list, 
                                                       stats    = rank,
                                                       minSize  = 8,
                                                       maxSize  = 500, 
                                                       eps = 0, 
                                                       BPPARAM = MulticoreParam(workers = core_n - 4)) %>% 
                            as_tibble()
}))

gsea_res <- gsea_exp %>% select(contrast, gs_name, project, gse_res) %>% unnest()
# gsea_res %>% saveRDS("data/gsea_res.rds")
```








