# Skin transcriptome data cleaning

## Baseline characteristics

```{r baseline-statistics-table}
extensive_meta <- 
  readr::read_csv("../multiomics-ad-transcriptomics/data/RNAseq_sample_annotation(extensive).csv") 

baseline_data <-  
  extensive_meta %>% filter(visit == "01") %>% 
  select(subject, 
         group, 
         # t test
         age, 
         bmi, 
         contains("blood"),
         # fisher's exact test
         gender, 
         smoker,
         smoker_ever,
         animal,
         family_eczema,
         family_atopy,
         other_atopy,
         other_diagnoses,
         prick_test_result,
         # Mann-Whitney-Wilcoxon Test
         education,
         drinks_per_week,
         skin_type_fitzpatrick_scale,
         grow_up_where,
         # NOT FOR COMPARISON
         easi_total_score, scorad, 
         scorad_objective) %>% 
  distinct() %>% 
  mutate(ige_200h = ifelse(blood_ige > 200, TRUE, FALSE),
         gender_female = ifelse(gender == "female", T, F),
         family_eczema = ifelse(family_eczema == "-", 
                                F,
                                family_eczema),
         family_eczema = as.logical(family_eczema),
         other_atopy = ifelse(other_atopy == "-",
                              F,
                              other_atopy),
         other_atopy = as.logical(other_atopy),
         prick_test_result = case_when(prick_test_result == "POS" ~ TRUE,
                                       prick_test_result == "NEG" ~ FALSE,
                                       TRUE ~ NA),
         education = ifelse(education == "meidium_higher_education",
                            "medium_higher_education",
                            education),
         education = case_when(education == "primary_school" ~ 0,
                               education == "high_school" ~ 1,
                               education == "short_higher_education" ~ 2,
                               education == "medium_higher_education" ~ 3,
                               education == "long_higher_education" ~ 4),
         drinks_per_week = case_when(drinks_per_week == "0" ~ 0,
                                     drinks_per_week == "1_7" ~ 1,
                                     drinks_per_week == "7_14" ~ 2,
                                     drinks_per_week == "15_" ~ 3))

# Mann-Whitney-Wilcoxon Test
wilcox_test_d <-
  baseline_data %>% 
  select(subject, group, 
         education, skin_type_fitzpatrick_scale, drinks_per_week) %>% 
  pivot_longer(cols = c("education", "skin_type_fitzpatrick_scale", "drinks_per_week"),
               names_to = "parameter") %>% 
  group_by(parameter) %>% nest() %>% 
  mutate(p_value = map_dbl(data, function(data){
    CO <- data %>% filter(group == "CO")
    AD <- data %>% filter(group == "AD")
    wilcox.test(CO$value, AD$value, exact = F)$p.value
  }))

# t test
t_test_d <- 
  baseline_data %>% 
  select(subject, group, contains("blood"), age, bmi) %>% 
  pivot_longer(cols = !c("subject", "group"), 
               names_to = "parameter") %>% 
  group_by(parameter) %>% nest() %>% 
  mutate(p_value = map(data, function(data){
    CO <- data %>% filter(group == "CO")
    AD <- data %>% filter(group == "AD")
    t.test(CO$value, AD$value, na.rm = T)$p.value
  }))


# fisher's exact test
fisher_test_d <-
  baseline_data %>% 
  select(subject, group, gender_female, smoker, smoker_ever,
         animal, family_eczema, family_atopy, 
         other_atopy, other_diagnoses,
         prick_test_result) %>% 
  pivot_longer(cols = !c("subject", "group"),
               names_to = "parameter") %>% 
  group_by(parameter) %>% nest() %>% 
  mutate(
    AD = map_dbl(data, function(data){
      data %>% filter(group == "AD", value == TRUE) %>% nrow()
    }),
    CO = map_dbl(data, function(data){
      data %>% filter(group == "CO", value == TRUE) %>% nrow()
    }),
    p_value = map_dbl(data, function(data){
      test <- data %>% select(-subject) %>% table %>% fisher.test()
      test$p.value
      })
    )


  
clinic_summary <- 
 baseline_data %>% select(-subject) %>% group_by(group) %>% 
  summarize(no_subject = n(),
            female_percent = mean(gender == "female") * 100,
            age_mean = mean(age), age_sd = sd(age), 
            bmi_mean = mean(bmi), bmi_sd = sd(bmi), 
            blood_ige_mean = mean(blood_ige, na.rm = TRUE), blood_ige_sd = sd(blood_ige, na.rm = TRUE),
            blood_leukocyte_mean = mean(blood_leukocyte, na.rm=T), 
            blood_leukocyte_sd = sd(blood_leukocyte, na.rm=T),
            blood_lymphocyte_mean = mean(blood_lymphocyte, na.rm=T), 
            blood_lymphocyte_sd = sd(blood_lymphocyte, na.rm=T),
            blood_monocyte_mean = mean(blood_monocyte, na.rm=T),
            blood_monocyte_sd = sd(blood_monocyte, na.rm=T),
            blood_neutrophil_mean = mean(blood_neutrophil, na.rm = T),
            blood_neutrophil_sd = sd(blood_neutrophil, na.rm = T),
            blood_eosinophil_mean = mean(blood_eosinophil, na.rm = T),
            blood_eosinophil_sd = sd(blood_eosinophil, na.rm = T),
            blood_basophil_mean = mean(blood_basophil, na.rm = T),
            blood_basophil_sd = sd(blood_basophil, na.rm = T),
            easi_mean = mean(easi_total_score), easi_sd = sd(easi_total_score), 
            scorad_mean = mean(scorad, na.rm = TRUE), scorad_sd = sd(scorad, na.rm = TRUE),
            oscorad_mean = mean(scorad_objective, na.rm = TRUE), 
            oscorad_sd = sd(scorad_objective, na.rm = TRUE),
            ige_mean = mean(blood_ige, na.rm = TRUE), 
            ige_sd = sd(blood_ige, na.rm = TRUE),
            ige_median = median(blood_ige, na.rm = TRUE),
            ige_200h = sum(blood_ige > 200, na.rm = TRUE))

p_value <- 
  tibble(
    `Female%` = baseline_data %>% select(gender, group) %>% 
      table %>% fisher.test() %>% broom::tidy() %>% pull(p.value),
    `Blood_Ige > 200` = baseline_data %>% select(group, ige_200h) %>% 
      table %>% fisher.test() %>% broom::tidy() %>% pull(p.value),
    age = t.test(x = baseline_data %>% filter(group == "AD") %>% select(age),
                 y = baseline_data %>% filter(group == "CO") %>% select(age))$p.value,
    bmi = t.test(x = baseline_data %>% filter(group == "AD") %>% select(bmi),
                 y = baseline_data %>% filter(group == "CO") %>% select(bmi))$p.value,
    `blood_ige` = t.test(x = baseline_data %>% filter(group == "AD") %>% select(blood_ige),
                         y = baseline_data %>% filter(group == "CO") %>% select(blood_ige))$p.value) %>% 
  mutate_all(round, digits = 2) %>% mutate_all(as.character)

clinic_summary_t <-
  clinic_summary %>% 
  mutate(group = paste0(group, " (n=", no_subject,")"),
         `Female%` = female_percent %>% round(),
         age = paste(age_mean %>% round(1) , "±", age_sd %>% round(1)),
         bmi = paste(bmi_mean %>% round(1) , "±", bmi_sd %>% round(1)),
         blood_ige = paste(blood_ige_mean %>% round(1) , "±", blood_ige_sd %>% round(1)),
         `Blood_Ige > 200` = round(ige_200h / no_subject, 2) * 100,
         easi = paste(easi_mean %>% round(1) , "±", easi_sd %>% round(1)),
         scorad = paste(scorad_mean %>% round(1) , "±", scorad_sd %>% round(1)),
         oscorad = paste(oscorad_mean %>% round(1) , "±", oscorad_sd %>% round(1))) %>%
  select(group, `Female%`:oscorad) %>%
  column_to_rownames("group") %>% mutate_all(as.character) %>% 
  add_row(tibble_row(p_value)) %>% 
  t()

colnames(clinic_summary_t)[3] <- "p-value"
clinic_summary_t[6:8, 2:3] <- " / "
rownames(clinic_summary_t)[1] <- "Female(%)"
rownames(clinic_summary_t)[2] <- "Age"
rownames(clinic_summary_t)[3] <- "BMI"
rownames(clinic_summary_t)[4] <- "Blood IgE"
rownames(clinic_summary_t)[5] <- "Blood IgE>200(%)"
rownames(clinic_summary_t)[6] <- "EASI"
rownames(clinic_summary_t)[7] <- "SCORAD"
rownames(clinic_summary_t)[8] <- "oSCORAD"
colnames(clinic_summary_t)[2] <- "HC (n=30)"
knitr::kable(clinic_summary_t, 
             format = "html", 
             caption = "Baseline characteristics of AD and HC group") %>% 
  kableExtra::add_footnote("Age, bmi, and blood IgE were tested by t test. Gender and blood IgE > 200 (%) were tested by Fisher's exact test.")

clinic_summary_t_xlsx <- clinic_summary_t %>% as_tibble(rownames = " ")
clinic_summary_t_xlsx[9,1] <- "notes: Age, bmi, and blood IgE were tested by t test. Gender and blood IgE > 200 (%) were tested by Fisher's exact test."
# openxlsx::write.xlsx(clinic_summary_t_xlsx, 
#                      "data/supplementary/table_s1.xlsx", overwrite = T)
```


## Differential expressed genes: baseline

```{r DGE-V1, cache=TRUE}
# V1 LSvsNL --------------------
se_v1_LSvsNL <- se[, se$visit == "01" & se$skin_type %in% c("LS", "NL")]
de_v1_LSvsNL <- DESeq2::DESeqDataSet(se_v1_LSvsNL, ~ subject + skin_type)
de_v1_LSvsNL <- DESeq2::DESeq(de_v1_LSvsNL, parallel = TRUE)
de_v1_LSvsNL_LFC <- lfcShrink(de_v1_LSvsNL, coef = "skin_type_LS_vs_NL", type = "apeglm", parallel = TRUE, quiet = TRUE)
de_v1_LSvsNL_result <- de_v1_LSvsNL_LFC %>% as.data.frame() %>% tibble::rownames_to_column(var = "gene_name") %>% tibble()
de_v1_LSvsNL_result_s <- de_v1_LSvsNL_result %>% filter(padj < .05, abs(log2FoldChange) > 1)

# V1 LSvsNN --------------------
se_v1_LSvsNN <- se[, se$visit == "01" & se$skin_type %in% c("LS", "NN")]
de_v1_LSvsNN <- DESeq2::DESeqDataSet(se_v1_LSvsNN, ~ gender + skin_type) 
de_v1_LSvsNN <- DESeq2::DESeq(de_v1_LSvsNN, parallel = TRUE)
de_v1_LSvsNN_LFC <- lfcShrink(de_v1_LSvsNN, coef = "skin_type_LS_vs_NN", type = "apeglm", parallel = TRUE, quiet = TRUE)
de_v1_LSvsNN_result <- de_v1_LSvsNN_LFC %>% as.data.frame() %>% tibble::rownames_to_column(var = "gene_name") %>% tibble()
de_v1_LSvsNN_result_s <- de_v1_LSvsNN_result %>% filter(padj < .05, abs(log2FoldChange) > 1)

# V1 LSvsNN --------------------
se_v1_NLvsNN <- se[, se$visit == "01" & se$skin_type %in% c("NL", "NN")]
de_v1_NLvsNN <- DESeq2::DESeqDataSet(se_v1_NLvsNN, ~ gender + skin_type) 
de_v1_NLvsNN <- DESeq2::DESeq(de_v1_NLvsNN, parallel = TRUE)
de_v1_NLvsNN_LFC <- lfcShrink(de_v1_NLvsNN, coef = "skin_type_NL_vs_NN", type = "apeglm", parallel = TRUE, quiet = TRUE)
de_v1_NLvsNN_result <- de_v1_NLvsNN_LFC %>% as.data.frame() %>% tibble::rownames_to_column(var = "gene_name") %>% tibble()
de_v1_NLvsNN_result_s <- de_v1_NLvsNN_result %>% filter(padj < .05, abs(log2FoldChange) > 1)

# Consolidate the results -----
de_v1_c <-
  bind_rows(de_v1_LSvsNL_result_s %>% mutate(contrast = "LSvsNL"),
            de_v1_LSvsNN_result_s %>% mutate(contrast = "LSvsNN"),
            de_v1_NLvsNN_result_s %>% mutate(contrast = "NLvsNN")) %>% 
  mutate(direction = ifelse(log2FoldChange > 0, "up", "down"))
de_v1_c_s <-
  de_v1_c %>% 
  group_by(contrast, direction) %>% 
  summarise(n_dys = n())
# No of dysregulated genes
n_de_v1_LSvsNL <- de_v1_c_s %>% filter(contrast == "LSvsNL") %>% pull(n_dys)
n_de_v1_LSvsNN <- de_v1_c_s %>% filter(contrast == "LSvsNN") %>% pull(n_dys)
n_de_v1_NLvsNN <- de_v1_c_s %>% filter(contrast == "NLvsNN") %>% pull(n_dys)
de_v1_c %>% readr::write_csv("data/de_v1_sig.csv")
de_v1_full <- list(de_v1_LSvsNL = de_v1_LSvsNL_result, 
                   de_v1_LSvsNN = de_v1_LSvsNN_result, 
                   de_v1_NLvsNN = de_v1_NLvsNN_result)
# de_v1_full %>% readr::write_rds("data/de_v1_full.rds")
```

## Differential expressed genes: anotomic area
```{r DGE space, cache=TRUE, eval=FALSE}
DGE_space <-
  tibble(
    C1 = c("LS", "NN"),
    C2 = c("NL", "NN")
  ) %>% 
  mutate(
    contrast = paste0(C1, "vs", C2),
    se_exp = map2(C1, C2, ~ se[,se$skin_type %in% c(.x, .y)]),
    design_f = ifelse(C1 == "NN" & C2 == "NN", 
                      "~ subject + visit + biopsy_area", 
                      "~ subject + skin_type + visit + biopsy_area"),
    deseq = map2(se_exp, design_f, ~ DESeqDataSet(.x, .y %>% as.formula)),
    deseq = map(deseq, ~ DESeq(.x, parallel = T)),
    res_name = map(deseq, ~ resultsNames(.x) %>% grep(pattern = "biopsy_area", ., value = T))) %>% 
  unnest(cols = res_name) %>% 
  mutate(
    res_shrink = map2(deseq, res_name, ~ lfcShrink(.x, coef = .y, type = "apeglm", parallel = T)),
    res_shrink = map(res_shrink, ~ as_tibble(.x, rownames = "gene_name"))
  )
save(DGE_space %>% select(contrast, res_shrink), "data/dge_space.rds")
```

## Differential expressed genes: time

```{r load and clean data}
se <- readr::read_rds("data/se.rds")
se <- se[,!is.na(se$visit_quarter)]
```

### Time stability (dge_tstat)

```{r dge_time_alone, eval=FALSE}
dge_tstat_design <-
  expand_grid(
  skin_type = colData(se)$skin_type %>% unique,
  t = c("visit", "visit_quarter")
  ) %>% 
  mutate(name = paste(skin_type, t, sep = "_"))
  
dge_tstat_result <-
  plyr::llply(1:nrow(dge_tstat_design), function(index){
    se <- se[,se$skin_type == dge_tstat_design$skin_type[index]]
    f <- paste("~", dge_tstat_design$t[index])
    deseq <- DESeqDataSet(se, design = as.formula(f))
    deseq <- DESeq(deseq, test = "LRT", reduced = ~ 1, 
                   parallel = T, 
                   BPPARAM = MulticoreParam(min(detectCores() - 2, 8)))
    deseq <- results(deseq)
    deseq <- as_tibble(deseq, rownames = "gene_name") %>% 
      mutate(name = dge_tstat_design$name[[index]])
  }, .progress = "text")
# saveRDS(dge_tstat_result, "data/dge_tstat_result.rds")
# dge_tstat_result <- readr::read_rds("data/dge_tstat_result.rds")

dge_tstat_result_filt <-
  map_dfr(dge_tstat_result,
           ~ filter(.x, padj < .05, abs(log2FoldChange) > 1)) 

library(enrichR)
setEnrichrSite("Enrichr")

dge_tstat_result_filt_n <- 
  dge_tstat_result_filt %>% group_by(name) %>% nest() %>% 
  mutate(gene_name = map(data, ~ pull(.x, gene_name)),
         enrichment = map(gene_name, ~ enrichr(., c("Reactome_2016"))))

dge_tstat_g <- dge_tstat_result_filt$gene_name %>% unique
dbs <- c("Human_Gene_Atlas", "ARCHS4_Tissues")
enriched_dge_tstat <- enrichr(dge_tstat_g, dbs)

map2_dfr(enriched_dge_tstat, dbs, 
         ~ .x %>% mutate(name = .y) %>% 
           filter(Adjusted.P.value < .05))
```

### Disease signature stability
```{r load dge-time}
dge_time <- 
  readr::read_rds("data/dge_time.rds")

names(dge_time)[1:2] <- 
  c("DGE_condition_specific",
    "DGE_condition_specific_non")

dge_time[[2]] %>% unnest(cols = "result") %>% 
  filter(padj < .05, abs(log2FoldChange) > 1) %>% 
  group_by(t, contrast) %>% 
  summarise(n = n())

ad_sig_stat_g <- 
  dge_time[[2]] %>% unnest(cols = "result") %>% 
    filter(padj < .05, abs(log2FoldChange) > 1) %>% filter(contrast =="LSvsNL", t =="visit_quarter") %>% pull(gene_name)

```

### Condition(LS/NL/HC)-specific change
```{r}
dge_time[[1]] %>% unnest(cols = "result") %>% 
  filter(padj < .05, abs(log2FoldChange) > 1) %>% 
  group_by(t, contrast) %>% summarise(n = n())

condition_specific_g <- 
  dge_time[[1]] %>% unnest(cols = "result") %>% 
  filter(padj < .05, abs(log2FoldChange) > 1) %>% 
  group_by(t, contrast) %>% nest() 

se[c("SAA2", "KRT27", "MUC1", "MGST1",
     "KRT71", "SHISA2", "PTGS2"),] %>% as_tibble() %>% 
    ggline(x = "visit_quarter", 
           y = "counts_scaled",
           facet.by = "feature",
           add = c("mean", "dotplot"),
           color = "skin_type", 
           palette = pca_color,
           xlab = "Quarter",
           ylab = "Counts") %>% 
    ggpar(yscale = "log10")

se[c("KRT16", "KRT18", "KRT27", 
     "KRT7", "KRT71", "KRT80"),] %>% as_tibble() %>% 
    ggline(x = "visit_quarter", 
           y = "counts_scaled",
           facet.by = "feature",
           add = c("mean", "dotplot"),
           color = "skin_type", 
           palette = pca_color,
           xlab = "Quarter",
           ylab = "Counts") %>% 
    ggpar(yscale = "log10")

```

### Time outlier figure
```{r}
time_outlier <-
  dge_tstat_result_filt %>% 
  select(gene_name, name) %>% 
  mutate(skin_type = str_extract(name, "(LS|NL|NN)"),
         skin_type = ifelse(skin_type == "NN", "HC", skin_type),
         t = str_extract(name, "(visit$|visit_quarter)")) %>% 
  select(-name) %>% 
  group_by(skin_type, t) %>% nest() %>% 
  mutate(g_n = map(data, "gene_name")) %>% 
  mutate(figure = 
           pmap(list(g_n, skin_type, t),
                       function(g_n, skin_type, t){
  plot <-
    se[g_n, se$skin_type == skin_type] %>% as_tibble() %>% 
    ggboxplot(x = t,
              y = "counts_scaled",
              facet.by = "feature",
              xlab = ifelse(t == "visit_quarter", 
                            "quarter",
                            t),
              ylab = FALSE,
              scales = "free", title = skin_type)
  }),
        save_f = pmap(list(skin_type, t, figure),
                      function(skin_type, t, figure){
          name = paste0("supplement_t_outlier_", paste(skin_type, t, sep = "_"),".png")
          ggsave(filename = paste0("figure/supplement_t_outlier", name),
                 figure, width = 16.8, height = 11.8)
        }))

```

## Biological replicates
```{r found biological replicates}

bioreplicates <-
  colData(se) %>% as_tibble() %>% 
  mutate(biological_rep_id = paste(subject, visit, skin_type)) %>%
  group_by(biological_rep_id) %>% 
  summarise(n = n()) %>% 
  filter(n == 2) %>% 
  pull(biological_rep_id)

bioreplicates_t <- 
  colData(se) %>% as_tibble() %>% 
  mutate(biological_rep_id = paste(subject, visit, skin_type)) %>% 
  filter(biological_rep_id %in% bioreplicates) %>% 
  left_join(se[c("CCL17", "CCL18", "CCL20", "CCL22", "CCL26",
                 "CXCL1", "CXCL10", "CXCL9", "FOXP3",
                 "IL10", "IL22", "IL34", "IL37", "IL4R",
                 "LORICRIN", "MMP12", "OSMR", "PI3",
                 "S100A12", "S100A7", "STAT1"),] %>% as_tibble())


replicate_g <-
  bioreplicates_t %>% 
    select(biological_rep_id, 
           subject, visit, skin_type, feature, 
           counts_scaled, replicate_ID) %>% 
  ggplot(aes(x = replicate_ID, 
             y = counts_scaled, 
             group = biological_rep_id)) + 
  geom_point(aes(color = skin_type)) + 
  facet_grid(feature ~ skin_type, scales = "free") + 
  geom_line(aes(color = skin_type), alpha = .5) + 
  geom_boxplot(aes(group = replicate_ID, fill = skin_type), alpha = .5) + 
  scale_color_manual(values = pca_color) + 
  scale_fill_manual(values = pca_color) + 
  scale_y_continuous(trans = "log2") +
  xlab("replicate") +
  theme(axis.title.y = element_blank())

ggsave("data/supplementary/supplement_replicate.png", replicate_g,
       height = 25, width = 8)
```

### correlation coefficient
```{r}
ad_sig <- 
  readr::read_csv("data/supplementary/table_s2.csv",
                  show_col_types = F) %>% 
  pull(gene_name) %>% 
  unique

bioreplicates_sig_t <- 
  colData(se) %>% as_tibble() %>% 
  mutate(biological_rep_id = paste(subject, visit, skin_type)) %>% 
  filter(biological_rep_id %in% bioreplicates) %>% 
  left_join(se[ad_sig,] %>% as_tibble())
```




## Gene set enrichment analysis
```{r dge result}
de_v1_full <- readr::read_rds("data/de_v1_full.rds")
gse107361 <- 
  readr::read_rds("data/dge_gse107361.rds") %>% # adult vs children
  mutate(dge_result = map(dge_result, unnest))
  
gse120721 <- readr::read_rds("data/dge_gse120721.rds") # LSM
gse121212 <- readr::read_rds("data/dge_gse121212.rds")
ChronicvsAcute <- readr::read_rds("data/dge_AcutevsChronic.rds")

dge_result <-
  tibble(
    project = c(rep("multi-omics",length(de_v1_full)), 
                rep("gse121212",nrow(gse121212)),
                rep("gse107361",nrow(gse107361)),
                rep("gse120721",nrow(gse120721)),
                rep("acute_chronic", nrow(ChronicvsAcute))),
    contrast = c(de_v1_full %>% names %>% str_remove("de_v1_"), 
                 gse121212$contrast, 
                 gse107361$contrast,
                 gse120721$contrast,
                 ChronicvsAcute$contrast),
    dge_result = c(de_v1_full, 
                   gse121212$res_shrink,
                   gse107361$dge_result,
                   gse120721$dge_result,
                   ChronicvsAcute$data))
```

```{r prepare gene set, cache=TRUE}
# prepare gs --------------
gs <- 
  tibble(gs_name = c("CP:BIOCARTA", 
                     "CP:KEGG", 
                     "CP:PID", 
                     "CP:REACTOME", 
                     "CP:WIKIPATHWAYS", 
                     "MIR:MIRDB", 
                     "TFT:GTRD", 
                     "GO:BP", 
                     "GO:MF")) %>% 
  mutate(gs_list = map(gs_name, function(subcat){
  gs_df <- msigdbr::msigdbr(species = "Homo sapiens", 
                            subcategory = subcat) %>% 
    group_by(gs_name) %>% tidyr::nest() %>% 
    mutate(gene_id = purrr::map(data, ~ .x %>% pull(gene_symbol))) %>% select(-data)
  gs_list <- gs_df$gene_id
  names(gs_list) <- gs_df$gs_name
  return(gs_list)
}))
```

```{r fgsea, cache=TRUE}
gsea_exp <- 
  expand.grid(contrast = dge_result$contrast,
              gs_name = gs$gs_name) %>% 
  left_join(dge_result, by = "contrast") %>% 
  left_join(gs, by = "gs_name") %>% 
  mutate(rank = map2(dge_result, gs_name,
                     function(dge_result, gs_name){
                       p_cutoff <- ifelse(gs_name %in% c("MIR:MIRDB", 
                                                         "TFT:GTRD"),
                                          1, # .05
                                          1)
                       log2fc_cutoff <- ifelse(gs_name %in% c("MIR:MIRDB",
                                                        "TFT:GTRD"),
                                               0, # 1
                                               0)
                       dge_result %>% 
                         mutate(rank = log2FoldChange) %>% 
                         filter(padj < p_cutoff, abs(log2FoldChange) > log2fc_cutoff) %>% 
                         select(gene_name, rank) %>% deframe}),
         gse_res = map2(rank, gs_list,
                            function(rank, gs_list){
                              gsa_stat <- fgsea::fgsea(pathways = gs_list, 
                                                       stats    = rank,
                                                       minSize  = 8,
                                                       maxSize  = 500, 
                                                       eps = 0,
                                                       nPermSimple = 10000,
                                                       BPPARAM = MulticoreParam(workers = core_n - 4)) %>% 
                            as_tibble()
}))

gsea_res <- gsea_exp %>% select(contrast, gs_name, project, gse_res) %>% unnest()
# gsea_res %>% saveRDS("data/gsea_res_FIVE_nocutoff.rds")
```

## LINC 
```{r}
de_v1_full <- readr::read_rds("data/de_v1_full.rds")

de_v1_full_t <-
  de_v1_full %>% map2_dfr(names(de_v1_full),
  function(x1,x2){
    x1 <- x1 %>% mutate(contrast = x2)
    }) %>% 
  filter(padj < .05, abs(log2FoldChange) > 1)

LINC <- 
  de_v1_full_t %>% filter(gene_name %>% str_detect("LINC"))

LINC %>% write.csv("data/LINC.txt")


  de_v1_full_t$gene_name %>% grep("LINC", . , value = T) %>% 
  unique()

l_g <- de_v1_full_t$gene_name %>% unique %>% length
l_linc <- length(LINC)
l_linc / l_g 
```


