# Skin Transcriptome in Time and Space

## Abstract

## Introduction

## Results

```{r load package, echo=FALSE, message=FALSE, warning=FALSE}
pacman::p_load(tibble, dplyr, tidyr, readr, stringr, ggplot2, purrr, 
               tidybulk, tidySummarizedExperiment, 
               ComplexHeatmap, ggpubr,
               BiocParallel, Biobase, variancePartition,
               DESeq2, limma, edgeR, MuSiC, convert, 
               patchwork)
knitr::opts_chunk$set(message = FALSE, warning = FALSE, echo = FALSE)
pca_color <- c("LS" = "#eb2d0c", "NL" = "#eb740c", "NN" = "#91cf60")
core_n <- future::availableCores()
register(MulticoreParam(ifelse(core_n <= 8, core_n - 2, core_n -6)))
```

```{r load data}
clinical_records <- 
  readr::read_csv("../multiomics-ad-transcriptomics/data/LEO-AD-metadata.csv") %>% 
  mutate(id = id %>% str_replace("-", "_"),
         JOIN_ID = paste(id, paste0("0",visit_no), sep="_"))
se <- readr::read_rds("data/se.rds")
se_krtap <- readr::read_rds("data/se_krtap.rds")
```

```{r baseline characteristics}
baseline_disease_severity <- 
  clinical_records %>% filter(visit_no == 1, group == "AD") %>% 
  summarise(mean_scorad = mean(scorad, na.rm = TRUE) %>% round(1),
            mean_easi = mean(easi_total_score) %>% round(2))
```

```{r follow-up subjects}
follow_up <- 
  clinical_records %>% filter(visit_no != 1) %>% 
  group_by(group) %>% 
  select(id, group) %>% 
  distinct() %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = group, values_from = n)
```


### Baseline characteristics of patients with mild to moderate AD

The atopic dermatitis (AD) group were characterized by mild to moderate disease severity, with mean eczema area and severity index (EASI) `r baseline_disease_severity$mean_easi` and mean SCORing Atopic Dermatitis (SCORAD) `r baseline_disease_severity$mean_scorad`. At the baseline, the AD and healthy control (HC) group showed no statistical difference in gender, age, BMI, and total blood IgE (Table: \@ref(tab:baseline-statistics-table)). Individual data points were reported in the supplementary table \@ref(tab:individual-data-table).


### RNA quality outliers and Data curation
We performed transcriptomic profiling on 401 skin biopsies collected from lesional and non-lesional skin of 30 AD patients and site-matched skin from 30 healthy subjects. From each subject, biological replicates were obtained from different skin sites. 
For `r follow_up$AD` AD and `r follow_up$CO` healthy participants, the sampling was repeated at five subsequent visits every two months for up to one year.
After data curation and cleaning (details in Data QC and curation), 339 samples were left for further analysis.


### Transcriptome architecture and components of AD

```{r PCA}
PCA <- 
  se %>% 
  reduce_dimensions(method = "PCA", top = 350)
```

```{r PCA 3d, eval=FALSE}
attr(PCA, "internals")$PCA %>% 
  pca3d::pca3d(group = PCA$skin_type, col = pca_color, radius = 1.5, 
               shape = "s", 
               show.plane = FALSE)
# rgl::snapshot3d("figure/pca3d.png", fmt="png")
# rgl::rgl.postscript("figure/pca3d.pdf", fmt="pdf")
```

```{r PCA-3d-plot, fig.cap = "pca 3d plot", dpi=300, out.width='80%'}
knitr::include_graphics("figure/pca3d.png")
```

```{r cleaning}
rm(list = ls()[!ls() %in% c("se", "se_krtap", "pca_color")])
```

We obtained `r ncol(se)` skin transcriptomes after data cleaning. The top three principal components could separate the different tissue types (fig-\@ref(fig:PCA-3d-plot)). The heatmap showed AD lesional skins (LS) can be separated from AD non-lesional skins (NL) and healthy control skins (HC), but, NL and HC can't be separated (fig-\@ref(fig:heatmap-top-50)). LS tissues showed higher heterogeneity than NL and HC tissues (fig-\@ref(fig:distance)).

```{r heatmap-top-50, fig.cap = "Heatmap", fig.align='center', dpi = 300, fig.height=8}
se_heatmap <- se
core_genes <- 
  tibble(gene_name = readr::read_csv("data/de_v1_sig.csv") %>% 
  filter(contrast == "LSvsNL") %>% pull(gene_name)) %>% 
  pull(gene_name)
se_heatmap <- se_heatmap[core_genes[core_genes %in% rownames(se_heatmap)],]
heatmap_data <- 
  se_heatmap %>% 
  keep_variable(.abundance = counts_scale,
                top = 50)
heatmap_data_m <-
  scale(assay(heatmap_data, 2) %>% t() %>% log1p(), 
        center = TRUE, scale = TRUE) %>% t()
sample_site_color <- 
  colData(heatmap_data) %>% as_tibble() %>% pull(biopsy_area) %>% unique()
color <- RColorBrewer::brewer.pal(length(sample_site_color), "Pastel2")
names(color) <- sample_site_color
heatmap_annotation <- 
  HeatmapAnnotation(
    `Tissue type` = colData(heatmap_data) %>% as_tibble() %>% pull(skin_type), 
    `Sampling site` = colData(heatmap_data) %>% as_tibble() %>% pull(biopsy_area),
    col = list(`Tissue type` = pca_color,
               `Sampling site` = color),
    # simple_anno_size = unit(1, "cm"),
    annotation_legend_param = list(`Tissue type` = list(nrow = 1, 
                                                        # grid_height = unit(2, "cm"), 
                                                        # grid_width = unit(1, "cm"), 
                                                        # title_gp = gpar(fontsize = 16),
                                                        labels_gp = gpar(fontsize = 9)),
                                   `Sampling site` = list(nrow = 1,
                                                          # grid_height = unit(2, "cm"), 
                                                          # grid_width = unit(1, "cm"), 
                                                          # title_gp = gpar(fontsize = 16),
                                                          labels_gp = gpar(fontsize = 9))
                                   ))

heatmap <- ComplexHeatmap::Heatmap(heatmap_data_m,
                        name = "gene expression",
                        column_dend_reorder = TRUE,
                        # clustering_method_columns = "complete",
                        show_column_names = FALSE,
                        show_row_dend = FALSE,
                        row_names_gp = gpar(fontsize = min(10, 800 * 1.5 / dim(heatmap_data_m)[1])),
                        top_annotation = heatmap_annotation,
                        heatmap_legend_param = list(
                          title_gp = gpar(fontsize = 14),
                          labels_gp = gpar(fontsize = 12)
                          ))

draw(heatmap, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)

# poster 480*8.1.5 480*10*1.5
# png("figure/heatmap.png", width = 480*8, height = 480*10, res=300)
# draw(heatmap, heatmap_legend_side = "right",
#      annotation_legend_side = "top", merge_legend = FALSE)
# dev.off()
```


Differential gene expression analysis generated AD disease signature which consists of `r knitr::load_cache("DGE-V1", "n_de_v1_LSvsNL")` (LS vs NL), `r knitr::load_cache("DGE-V1", "n_de_v1_LSvsNN")` (LS vs HC), and `r knitr::load_cache("DGE-V1", "n_de_v1_NLvsNN")` (NL vs HC) differentially expressed genes (\@ref(ad-dge-v1)).

Gene set enrichment analysis (GSEA) revealed that the skin transcriptome was enriched for cytokine and chemokine mediated inflammation, abnormal skin differentiation and barrier pertubation. We also identified enrichment of transcription factor targets and microRNA targets, including TP73, MAML1 (related with NOTCH signaling), PSMB5, TRIP13, MIR6768_5P, MIR6833_3P, and MIR6873_3P targets ((Table: \@ref(tab:gsea-regulatory))).

## Variance analysis

### Disease
IL37 was the best explained cytokine family gene. IL37 showed a decrease trend in HC \> NL \> LS. We also observed a few long noncoding RNA (lincRNA) with high variance can be explained by AD (Fig 2).

### Individual


### Time Effect plays a minor role in transcriptome variation

Likelihood Ratio Test (LRT) results showed no genes reacted in the condition-specific manner over time, compared to baseline samples.

I first fit a `full model` (`~ subject + time + tissue_type + tissue_type * time`), then a `reduced model` (`~ subject + time + tissue_type`) by imposing `tissue_type * time` as the constraint.

### Hair variance

```{r generate heatmap for krtap, eval=FALSE}
KRTAP_filt_d <- 
  se_krtap %>% assay(2) %>% log1p() %>% t() %>% scale() %>% t()

sample_site_color <- 
  colData(se_krtap) %>% as_tibble() %>% pull(biopsy_area) %>% unique()

color <- RColorBrewer::brewer.pal(length(sample_site_color), "Set3")
names(color) <- sample_site_color

heatmap_annotation <- 
  HeatmapAnnotation(
    `Tissue type` = colData(se_krtap) %>% as_tibble() %>% pull(skin_type), 
    `Sampling site` = colData(se_krtap) %>% as_tibble() %>% pull(biopsy_area),
    col = list(`Tissue type` = pca_color,
               `Sampling site` = color),
    # simple_anno_size = unit(1, "cm"),
    annotation_legend_param = list(`Tissue type` = list(nrow = 1), 
                                                        # grid_height = unit(2, "cm"), 
                                                        # grid_width = unit(1, "cm"), 
                                                        # title_gp = gpar(fontsize = 16),
                                                        # labels_gp = gpar(fontsize = 16)),
                                   `Sampling site` = list(nrow = 2) 
                                                          # grid_height = unit(2, "cm"), 
                                                          # grid_width = unit(1, "cm"), 
                                                          # title_gp = gpar(fontsize = 16),
                                                          # labels_gp = gpar(fontsize = 16))
                                   ))

# grid_height = unit(1, "cm"), grid_width = unit(5, "mm")
                                               
heatmap <- ComplexHeatmap::Heatmap(KRTAP_filt_d,
                        name = "gene expression",
                        column_dend_reorder = TRUE,
                        # clustering_method_columns = "complete",
                        show_column_names = FALSE,
                        show_row_dend = FALSE,
                        row_names_gp = gpar(fontsize = min(12, 800 * 1.5 / dim(KRTAP_filt_d)[1])),
                        top_annotation = heatmap_annotation,
                        heatmap_legend_param = list(
                          # title_gp = gpar(fontsize = 14),
                          # labels_gp = gpar(fontsize = 12)
                          ))

draw(heatmap, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)


# png("figure/heatmap_KRTAP.png", width = 480*4.5, height = 480*3, res=300)
# draw(heatmap, heatmap_legend_side = "right",
#      annotation_legend_side = "top", merge_legend = FALSE)
# dev.off()
```

```{r eval=FALSE}
psych::pairs.panels(se_krtap %>% assay(2) %>% log1p() %>% t())
```

## Space effect

### Variance Partition

```{r vp_biopsy_area, cache=TRUE}
se_space <- se
se_space_500 <- se_space %>% keep_variable()
vp_space_500 <- variancePartition::fitExtractVarPartModel(
    assay(se_space_500, 2), 
    ~ (1|skin_type) + (1|subject) + (1|visit) + (1|gender) + (1|biopsy_area), 
    colData(se_space_500) %>% as_tibble()
    )
# readr::write_rds(se_space_500, "data/se_space_500.rds")
# readr::write_rds(vp_space_500, "data/vp_space_500.rds")
```

```{r vp_biopsy_area_vis}
variancePartition::plotVarPart(vp_space_500)
```

```{r}
top10_space_g <- 
  vp_space_500 %>% as_data_frame() %>% 
  mutate(gene_name = se_space_500 %>% rownames()) %>% 
  slice_max(biopsy_area, n = 10) %>% pull(gene_name)
```

```{r}
top10_space <- se_space_500 %>% filter(feature %in% top10_space_g)
top10_space %>% 
  ggplot(aes(biopsy_area, counts_scaled)) + geom_point() + geom_boxplot() +
  facet_wrap(~feature, scales = "free") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```



#### Bioreplicate

```{r bioreplicate}

```

## Cell Type Deconvolution




```{r vp - linc genes, dpi=600, fig.height=8}
# LINC dotplot
linc_vp <- 
  vp_core_t %>% 
  filter(gene_name %>% str_detect("^LINC")) %>% 
  arrange(-skin_type) %>% 
  pivot_longer(cols = !gene_name, names_to = "term") %>% 
  filter(term == "skin_type") %>% 
  mutate(var_explained = round(value * 100, 2)) %>% 
  slice_max(order_by = var_explained, n = 15)
linc_vp_P <- 
  ggpubr::ggdotchart(
  linc_vp, x = "gene_name", y = "var_explained",
  color = "#FC4E07",
  xlab = "",
  sorting = "descending",
  rotate = TRUE, 
  add = "segments",
  ggtheme = ggpubr::theme_pubr(),
  dot.size = 4,
  ylab = "Var% explained by atopic dermatitis tissue"
)

# LINC boxplot
top_linc <- linc_vp %>% pull(gene_name) %>% head(5)
top_linc_boxplot <-
  se[top_linc,] %>% as_tibble() %>% 
  mutate(feature = feature %>% forcats::fct_relevel(top_linc)) %>% 
  ggboxplot(x = "skin_type", y = "counts_scaled", add = "jitter", 
            facet.by = "feature", xlab = FALSE,
            ylab = FALSE, color = "skin_type", 
            palette = pca_color, legend.title = "",
            font.legend = c(14), nrow = 1, yscale = "log10") +
  stat_compare_means()

# aggregate
linc_vp_P / top_linc_boxplot
```





```{r disease severity}
top_scorad_g <- vp_core_t %>% slice_max(scorad, n = 10) %>% pull(gene_name)

# top_scorad_x <- 
  se_core[top_scorad_g, ] %>% 
  ggplot(aes(scorad, counts_scaled)) + geom_point() +
    facet_wrap("feature")
```


```{r analyze individual factor, eval=FALSE}
vp_core_t %>% 
  filter(skin_type > .05, subject > .05) %>% 
  mutate(skin_type_subject_ratio = skin_type / (subject + .01)) %>% 
  arrange(skin_type_subject_ratio)
```

```{r boxplot - individual, eval=FALSE}
boxplot_individual_d <- se %>% 
  filter(feature %in% c("DOK7", "PSG5", "IGHA2", "PDZD7", "IL34"))

boxplot_individual_p <-
  boxplot_individual_d %>% as_tibble() %>% 
  ggboxplot(x = "skin_type", y = "counts_scaled", add = "jitter", 
            facet.by = "feature", scale = "free", xlab = FALSE,
            ylab = FALSE, color = "skin_type", 
            palette = pca_color, legend.title = "",
            font.legend = c(14))

boxplot_individual_p
```

```{r eval=FALSE}
vp_container <- htmltools::withTags(table(
  class = 'display',
  thead(
    tr(
      th(rowspan = 2, 'Gene name'),
      th(colspan = 3, 'log2 Fold Change'),
      th(colspan = 3, 'Adjusted p-value'),
      th(rowspan = 2, 'Var% explained by AD tissue type')
    ),
    tr(
      lapply(rep(c('LSvsNL', 'LSvsNN', "NLvsNN"), 2), th)
    )
  )
))


datatable(vp_core, 
          colnames = c("gene name", "LSvsNL", "LSvsNN", "NLvsNN", "LSvsNL", "LSvsNN", "NLvsNN", "Var% explained by AD tissue type"), 
          container = vp_container, rownames = FALSE,
          options = list(pageLength = 20)) %>% 
  formatRound(vp_core %>% colnames %>% str_extract("log2FoldChange_\\w{6}") %>% purrr::discard(is.na)) %>% 
  formatSignif(vp_core %>% colnames %>% str_extract("padj_\\w{6}") %>% purrr::discard(is.na), digits = 3) %>% 
  formatPercentage("var_explained", 2)
```
