# Knowledge Base

## Gene Set Enrichment Analysis (BINF keynote speaking)

Good morning, everyone. Thanks for giving me the opportunity to  give a keynote speaking at our BINF day. 

Let's get started.

Today I'll give a talk about gene set analysis. The title is "Gene Set Analysis: Rethinking on its 20 years".

The earliest literature I can find about this technology can be traced back to 1999, in an old nature genetics article. In that articles, the authors named the method "category enrichment". 

I assume all of you knowing this method very well.

This type of analysis has evolved 20 years. So, today I'll rethink Pros/Cons of the technology. Alternative approach?

Omics data. 
(gene expression, x10000 genes). how to interprete the data? database e.g. gene cards to query one by one. not realistic.

Gene set.
A set of related genes - function / pathway.

Gene set is the key of this method.

This method has a lot of names.
GSEA

### Enrichment algorithm
Enrichment algorithms can be classfied in two categories. The first category is over-representation analysis; the second is functional class scoring method.

Over-representation analysis uses a list of non-ranked genes; while functional class scoring methods uses all genes without filtering.

#### Over-representation algorithm
This method is very simple. 2X2 table.

Pros:
- Fast to calculate (enrichR) / theoritical distribution /mathematical formula
- detect the most significant effect in the data

But this method has a very obvious shortcoming. 

#### Aggregating score algorithm
Rafael Irizarry is a prof. in biostatistics and the founder of bioconductor. 

GSEA one of the most citied bioinformatics tools 26000 citations. But it does not have an official tool published on Bioconductor.

modified version (fgsea).

unofficially supported R scipt on GitHub repo.

Now we may know the reason 

GSEA 2002 

Piano
consolidate all aggregating score method into one interface.


## Rethinking

### Can be improved

### Can't 