# Transcriptome appendix
```{r setup-transcriptome-appendix, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, echo = FALSE)
library(dplyr)
library(tibble)
library(tidyr)
library(SummarizedExperiment)
library(DT)
library(stringr)
library(ggpubr)
```

## Table S1 - Baseline characteristics

```{r baseline-statistics-table}
extensive_meta <- 
  readr::read_csv("../multiomics-ad-transcriptomics/data/RNAseq_sample_annotation(extensive).csv")

baseline_data <-  
  extensive_meta %>% filter(visit == "01") %>% 
  select(subject, age, bmi, blood_ige, gender, group, easi_total_score, scorad, scorad_objective) %>% 
  distinct() %>% 
  mutate(ige_200_h = ifelse(blood_ige > 200, TRUE, FALSE))
  
clinic_summary <- 
 baseline_data %>% select(-subject) %>% group_by(group) %>% 
  summarize(no_subject = n(),
            female_percent = mean(gender == "female") * 100,
            age_mean = mean(age), age_sd = sd(age), 
            bmi_mean = mean(bmi), bmi_sd = sd(bmi), 
            blood_ige_mean = mean(blood_ige, na.rm = TRUE), blood_ige_sd = sd(blood_ige, na.rm = TRUE), 
            easi_mean = mean(easi_total_score), easi_sd = sd(easi_total_score), 
            scorad_mean = mean(scorad, na.rm = TRUE), scorad_sd = sd(scorad, na.rm = TRUE),
            oscorad_mean = mean(scorad_objective, na.rm = TRUE), 
            oscorad_sd = sd(scorad_objective, na.rm = TRUE),
            ige_mean = mean(blood_ige, na.rm = TRUE), 
            ige_sd = sd(blood_ige, na.rm = TRUE),
            ige_median = median(blood_ige, na.rm = TRUE),
            ige_200_h = sum(blood_ige > 200, na.rm = TRUE))

p_value <- 
  tibble(
    `Female%` = baseline_data %>% select(gender, group) %>% table %>% fisher.test() %>% broom::tidy() %>% pull(p.value),
    `Blood_Ige > 200` = baseline_data %>% select(group, ige_200_h) %>% table %>% fisher.test() %>% broom::tidy() %>% pull(p.value),
    age = t.test(x = baseline_data %>% filter(group == "AD") %>% select(age),
                 y = baseline_data %>% filter(group == "CO") %>% select(age))$p.value,
    bmi = t.test(x = baseline_data %>% filter(group == "AD") %>% select(bmi),
                 y = baseline_data %>% filter(group == "CO") %>% select(bmi))$p.value,
    `blood_ige` = t.test(x = baseline_data %>% filter(group == "AD") %>% select(blood_ige),
                 y = baseline_data %>% filter(group == "CO") %>% select(blood_ige))$p.value
) %>% mutate_all(round, digits = 2) %>% mutate_all(as.character)

clinic_summary_t <-
  clinic_summary %>% 
  mutate(group = paste0(group, " (n=", no_subject,")"),
         `Female%` = female_percent %>% round(),
         age = paste(age_mean %>% round(1) , "±", age_sd %>% round(1)),
         bmi = paste(bmi_mean %>% round(1) , "±", bmi_sd %>% round(1)),
         blood_ige = paste(blood_ige_mean %>% round(1) , "±", blood_ige_sd %>% round(1)),
         `Blood_Ige > 200` = round(ige_200_h / no_subject, 2) * 100,
         easi = paste(easi_mean %>% round(1) , "±", easi_sd %>% round(1)),
         scorad = paste(scorad_mean %>% round(1) , "±", scorad_sd %>% round(1)),
         oscorad = paste(oscorad_mean %>% round(1) , "±", oscorad_sd %>% round(1))) %>%
  select(group, `Female%`:oscorad) %>%
  column_to_rownames("group") %>% mutate_all(as.character) %>% 
  add_row(tibble_row(p_value)) %>% 
  t()

colnames(clinic_summary_t)[3] <- "p-value"
clinic_summary_t[6:8, 2:3] <- " / "
rownames(clinic_summary_t)[1] <- "Female(%)"
rownames(clinic_summary_t)[2] <- "Age"
rownames(clinic_summary_t)[3] <- "BMI"
rownames(clinic_summary_t)[4] <- "Blood IgE"
rownames(clinic_summary_t)[5] <- "Blood IgE>200(%)"
rownames(clinic_summary_t)[6] <- "EASI"
rownames(clinic_summary_t)[7] <- "SCORAD"
rownames(clinic_summary_t)[8] <- "oSCORAD"
colnames(clinic_summary_t)[2] <- "HC (n=30)"
knitr::kable(clinic_summary_t, 
             format = "html", 
             caption = "Baseline characteristics of AD and HC group") %>% 
  kableExtra::add_footnote("Age, bmi, and blood IgE were tested by t test. Gender and blood IgE > 200 (%) were tested by Fisher's exact test.")

clinic_summary_t_xlsx <- clinic_summary_t %>% as_tibble(rownames = " ")
clinic_summary_t_xlsx[9,1] <- "notes: Age, bmi, and blood IgE were tested by t test. Gender and blood IgE > 200 (%) were tested by Fisher's exact test."
openxlsx::write.xlsx(clinic_summary_t_xlsx, 
                     "data/supplementary/table_s1.xlsx", overwrite = T)
```

## Table S2 - individual data table
```{r individual-data-table}
individual_data <- 
  baseline_data %>% arrange(subject) %>% select(-ige_200_h) %>% 
  mutate(age = age %>% round,
         bmi = bmi %>% round(1))
colnames(individual_data)[3] <- "BMI"
colnames(individual_data)[4] <- "blood_IgE"
colnames(individual_data)[7] <- "EASI"
colnames(individual_data)[8] <- "SCORAD"
colnames(individual_data)[9] <- "oSCORAD"
knitr::kable(individual_data, caption = "Baseline statistics of individual subjects")
openxlsx::write.xlsx(individual_data, 
                     "data/supplementary/table_s2.xlsx", overwrite = T)
```

## Space
```{r load dge object}
dge_space <- readr::read_rds("data/dge_space.rds")
dge_space_filt <- 
  dge_space %>% unnest() %>% filter(padj < .05, abs(log2FoldChange) > 1)
readr::write_csv(dge_space_filt, "data/supplementary/dge_space.csv")
```


## Tissue type heterogenity
```{r distance, cache=TRUE, fig.cap="Cosine distance of three tissue types"}
se <- readr::read_rds("data/se.rds")
group_color <- c("LS" = "#eb2d0c", "NL" = "#eb740c", "NN" = "#91cf60")
cosine_st <-
    lapply(c("LS", "NL", "NN"), function(x){
    se %>% 
      subset(select = skin_type == x) %>% 
      assay(2) %>% 
      coop::cosine() %>% 
      as.numeric()})

names(cosine_st) <- c("LS", "NL", "NN")

cosine_df <- plyr::ldply(cosine_st, tibble, .name_repair = ~ c("cos_dist"),
                       .id = "skin_type") %>% tibble

cosine_df %>% 
  ggpubr::ggdensity(x = "cos_dist", fill = "skin_type", 
                    palette = group_color, xlab = "Cosine similarity") %>% 
  ggpubr::ggpar(legend.title = "tissue type")
```

## Visit 1: differentially expressed genes {#ad-dge-v1}
```{r ad-dge-v1}
ad_de_v1 <- readr::read_csv("data/de_v1_sig.csv") %>% select(-direction)

de_v1_c_d <- ad_de_v1 %>% 
  select(gene_name, log2FoldChange, padj, contrast) %>% 
  pivot_wider(names_from = contrast, values_from = c("log2FoldChange", "padj")) %>% 
  arrange(-log2FoldChange_LSvsNL)

t_container <- htmltools::withTags(table(
  class = 'display',
  thead(
    tr(
      th(rowspan = 2, 'Gene name'),
      th(colspan = 3, 'log2 Fold Change'),
      th(colspan = 3, 'Adjusted p-value')
    ),
    tr(
      lapply(rep(c('LSvsNL', 'LSvsNN', "NLvsNN"), 2), th)
    )
  )
))
  
datatable(de_v1_c_d, 
          colnames = c("gene name", "LSvsNL", "LSvsNN", "NLvsNN", "LSvsNL", "LSvsNN", "NLvsNN"), container = t_container, rownames = FALSE,
          options = list(pageLength = 20)) %>% 
  formatRound(de_v1_c_d %>% colnames %>% str_extract("log2FoldChange_\\w{6}") %>% purrr::discard(is.na)) %>% 
  formatSignif(de_v1_c_d %>% colnames %>% str_extract("padj_\\w{6}") %>% purrr::discard(is.na), digits = 3)
```

## Visit 1: gene set enrichment analysis
```{r load gsea results}
gse <- readr::read_rds("data/gse_res.rds")
names(gse)
```

### Enrichment of AD disease signatures
```{r ad-signature, fig.width=8}
gse_df <- gse %>% purrr::map_dfr(~.x) %>% 
  filter(pathway %in% c("KEGG_CHEMOKINE_SIGNALING_PATHWAY",
                        "KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY",
                        "KEGG_JAK_STAT_SIGNALING_PATHWAY",
                        "KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY",
                        "GOBP_KERATINOCYTE_DIFFERENTIATION")) %>% 
  mutate(log10p = -log10(padj))

gse_df %>% 
  ggbarplot(x = "pathway", y = "NES", ylab = "Normalized enrichment score", xlab = "") %>% 
  facet("contrast", nrow = 2) %>% 
  ggpar(rotate = TRUE)

```


### Transcription factor and microRNA targets
```{r gsea-regulatory}
gse_reg <- 
  gse[grep("MIR|TFT", names(gse), value=T)] %>% purrr::map_dfr(~.x) %>% 
  bind_rows(gse[grep("CP:PID", names(gse), value =T)] %>% purrr::map_dfr(~.x) %>% 
              filter(pathway == "PID_P73PATHWAY"))


gse_reg %>% 
  select(gene_set = pathway, padj, NES, contrast) %>% 
  arrange(contrast) %>% 
  mutate(padj = signif(padj, digits = 2), 
         NES = round(NES, digits = 2),
         link = paste0("https://www.gsea-msigdb.org/gsea/msigdb/cards/", gene_set, ".html")) %>% 
  knitr::kable(caption = "Enrichment results for transcription factor and microRNA targets")
```




