# Transcriptome appendix
```{r setup-transcriptome-appendix, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, echo = FALSE)
library(dplyr)
library(tibble)
library(tidyr)
library(SummarizedExperiment)
library(DT)
library(stringr)
library(ggpubr)
pca_color <- c("LS" = "#eb2d0c", "NL" = "#eb8b9b", "HC" = "#91cf60")
```


## Table S2 - individual data table
```{r individual-data-table}
individual_data <- 
  baseline_data %>% arrange(subject) %>% select(-ige_200_h) %>% 
  mutate(age = age %>% round,
         bmi = bmi %>% round(1))
colnames(individual_data)[3] <- "BMI"
colnames(individual_data)[4] <- "blood_IgE"
colnames(individual_data)[7] <- "EASI"
colnames(individual_data)[8] <- "SCORAD"
colnames(individual_data)[9] <- "oSCORAD"
knitr::kable(individual_data, caption = "Baseline statistics of individual subjects")
openxlsx::write.xlsx(individual_data, 
                     "data/supplementary/table_s2.xlsx", overwrite = T)
```

## Space
```{r load dge object}
dge_space <- readr::read_rds("data/dge_space.rds")
dge_space_filt <- 
  dge_space %>% unnest() %>% filter(padj < .05, abs(log2FoldChange) > 1)
readr::write_csv(dge_space_filt, "data/supplementary/dge_space.csv")
```


## Tissue type heterogenity
```{r distance, cache=TRUE, fig.cap="Cosine distance of three tissue types"}
se <- readr::read_rds("data/se.rds")
group_color <- c("LS" = "#eb2d0c", "NL" = "#eb740c", "NN" = "#91cf60")
cosine_st <-
    lapply(c("LS", "NL", "NN"), function(x){
    se %>% 
      subset(select = skin_type == x) %>% 
      assay(2) %>% 
      coop::cosine() %>% 
      as.numeric()})

names(cosine_st) <- c("LS", "NL", "NN")

cosine_df <- plyr::ldply(cosine_st, tibble, .name_repair = ~ c("cos_dist"),
                       .id = "skin_type") %>% tibble

cosine_df %>% 
  ggpubr::ggdensity(x = "cos_dist", fill = "skin_type", 
                    palette = group_color, xlab = "Cosine similarity") %>% 
  ggpubr::ggpar(legend.title = "tissue type")
```

## Visit 1: differentially expressed genes {#ad-dge-v1}
```{r ad-dge-v1}
ad_de_v1 <- readr::read_csv("data/de_v1_sig.csv") %>% select(-direction)

de_v1_c_d <- ad_de_v1 %>% 
  select(gene_name, log2FoldChange, padj, contrast) %>% 
  pivot_wider(names_from = contrast, values_from = c("log2FoldChange", "padj")) %>% 
  arrange(-log2FoldChange_LSvsNL)

t_container <- htmltools::withTags(table(
  class = 'display',
  thead(
    tr(
      th(rowspan = 2, 'Gene name'),
      th(colspan = 3, 'log2 Fold Change'),
      th(colspan = 3, 'Adjusted p-value')
    ),
    tr(
      lapply(rep(c('LSvsNL', 'LSvsNN', "NLvsNN"), 2), th)
    )
  )
))
  
datatable(de_v1_c_d, 
          colnames = c("gene name", "LSvsNL", "LSvsNN", "NLvsNN", "LSvsNL", "LSvsNN", "NLvsNN"), container = t_container, rownames = FALSE,
          options = list(pageLength = 20)) %>% 
  formatRound(de_v1_c_d %>% colnames %>% str_extract("log2FoldChange_\\w{6}") %>% purrr::discard(is.na)) %>% 
  formatSignif(de_v1_c_d %>% colnames %>% str_extract("padj_\\w{6}") %>% purrr::discard(is.na), digits = 3)
```

## Visit 1: gene set enrichment analysis
```{r load gsea results}
gse <- readr::read_rds("data/gse_res.rds")
names(gse)
```

### Enrichment of AD disease signatures
```{r ad-signature, fig.width=8}
gse_df <- gse %>% purrr::map_dfr(~.x) %>% 
  filter(pathway %in% c("KEGG_CHEMOKINE_SIGNALING_PATHWAY",
                        "KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY",
                        "KEGG_JAK_STAT_SIGNALING_PATHWAY",
                        "KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY",
                        "GOBP_KERATINOCYTE_DIFFERENTIATION")) %>% 
  mutate(log10p = -log10(padj))

gse_df %>% 
  ggbarplot(x = "pathway", y = "NES", ylab = "Normalized enrichment score", xlab = "") %>% 
  facet("contrast", nrow = 2) %>% 
  ggpar(rotate = TRUE)

```


### Transcription factor and microRNA targets
```{r gsea-regulatory}
gse_reg <- 
  gse[grep("MIR|TFT", names(gse), value=T)] %>% purrr::map_dfr(~.x) %>% 
  bind_rows(gse[grep("CP:PID", names(gse), value =T)] %>% purrr::map_dfr(~.x) %>% 
              filter(pathway == "PID_P73PATHWAY"))


gse_reg %>% 
  select(gene_set = pathway, padj, NES, contrast) %>% 
  arrange(contrast) %>% 
  mutate(padj = signif(padj, digits = 2), 
         NES = round(NES, digits = 2),
         link = paste0("https://www.gsea-msigdb.org/gsea/msigdb/cards/", gene_set, ".html")) %>% 
  knitr::kable(caption = "Enrichment results for transcription factor and microRNA targets")
```

## Novel targes

### IL34 and IL37
```{r data - IL34, IL37, CSF1, CSF1R}
novel_target_expr <- 
  se[c("IL34", "IL37", "CSF1", "CSF1R")] %>% 
  select(feature, sample, counts_scaled, skin_type) %>% 
  pivot_wider(names_from = feature, values_from = counts_scaled)
```

```{r}
p_IL34_IL37_sep <- 
  novel_target_expr %>% 
  ggscatter(x = "IL34", y = "IL37", add = "reg.line", conf.int = T, color = "skin_type",
            cor.coef = T, cor.method = "spearman", 
            facet.by = "skin_type", scales = "free", palette = pca_color)

png("figure/p_IL34_IL37.png", width = 480 * 2.2 * 3, height = 480 * 1.5 * 2, pointsize = 18, res = 300)
p_IL34_IL37_sep + theme(plot.margin = margin(.1,.5,.1,.1,"cm"))
dev.off()
```

### IL34 and CSF1

```{r}
p_IL34_CSF1_sep <- 
  novel_target_expr %>% 
  ggscatter(x = "IL34", y = "CSF1", add = "reg.line", conf.int = T, color = "skin_type",
            cor.coef = T, cor.method = "spearman", 
            facet.by = "skin_type", scales = "free", palette = pca_color)
png("figure/p_IL34_CSF1.png", width = 480 * 2.2 * 3, height = 480 * 1.5 * 2, pointsize = 18, res = 300)
p_IL34_CSF1_sep + theme(plot.margin = margin(.1,.5,.1,.1,"cm"))
dev.off()
```

### IL34 and CSF1R

```{r}
p_IL34_CSF1R_sep <- 
  novel_target_expr %>% 
  ggscatter(x = "IL34", y = "CSF1R", add = "reg.line", conf.int = T, color = "skin_type",
            cor.coef = T, cor.method = "spearman", 
            facet.by = "skin_type", scales = "free", palette = pca_color)
png("figure/p_IL34_CSF1R.png", width = 480 * 2.2 * 3, height = 480 * 1.5 * 2, pointsize = 18, res = 300)
p_IL34_CSF1R_sep + theme(plot.margin = margin(.1,.5,.1,.1,"cm"))
dev.off()
```




