---
title: "Figure S2 (Genome regulatory elements)"
author: "Tu Hu"
date: "06/07/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Genome regulatory elements (Figure S2)

### Transcription factor enrichment (Figure S2a)
```{r gsea - regulatory elements}
# gene set regulatory
gsea_r_p <- 
  gsea_res_filt %>% filter(pathway %in% gs_r, 
                           !contrast %>% str_detect("PSO"),
                           padj < .05,
                           !contrast %in% c("Adult_LSvsAdult_NN",
                                            "Adult_NLvsAdult_NN",
                                            "Adult_NLvsAdult_NN",
                                            "Children_NLvsChildren_NN",
                                            "Children_LSvsChildren_NN",
                                            "Adult_LSvsChildren_LS",
                                            "Adult_NLvsChildren_NL",
                                            "LS/NN_derm",
                                            "LS/NN_epi",
                                            "LS/NN_FT",
                                            "NL/NN_derm",
                                            "NL/NN_FT")) %>%
  mutate(project = ifelse(project == "multi-omics",
                          "GENAD",
                          project),
         project = factor(project, levels = c("gse121212",
                                         "acute_chronic",
                                         "gse107361",
                                         "gse120721",
                                         "GENAD"))) %>% 
  ggplot(aes(contrast, des, size = `-log10p`, color = NES)) + 
  geom_point() +
  scale_size_area(max_size = 6) + 
  scale_color_gradient2(low="blue", mid = "white", high="red") +
  theme_bw() +
  facet_grid(~project, scales = "free", space = "fixed") +
  theme(axis.title = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1))
ggsave("data/supplementary/figure_s2_a.png", gsea_r_p, 
       width = 6.5,
       height = 3.5,
       dpi = 300)
```

```{r figure 5c boxplot - ncRNA}
toptopncRNA <- 
  vp_core_t %>% 
  filter(gene_name %>% str_detect("LINC"),
         skin_type > .05) %>% 
  mutate(group = gene_name %>% str_extract("LINC")) %>%
  group_by(group) %>% 
  slice_max(skin_type, n = 5) %>% pull(gene_name)

toptopncRNA_boxplot <-
  se[toptopncRNA,] %>% as_tibble() %>% 
  mutate(feature = feature %>% forcats::fct_relevel(toptopncRNA)) %>% 
  ggboxplot(x = "skin_type", y = "counts_scaled", add = "jitter", 
            facet.by = "feature", xlab = FALSE,
            ylab = FALSE, color = "skin_type", 
            palette = pca_color, legend.title = "",
            font.legend = c(14), 
            nrow = 1, 
            yscale = "log10")

ggsave("data/supplementary/figure_s2_b.png", 
       toptopncRNA_boxplot, 
       width = 6.5,
       height = 3.5,
       dpi = 300)
```

#### LINC 
```{r}
LINC_g <- 
  de_v1_c %>% pull(gene_name) %>% 
  grep("LINC", ., value = T) %>% unique

MIR_g <- 
  de_v1_c %>% pull(gene_name) %>% 
  grep("MIR", ., value = T) %>% unique
```

```{r heatmap-linc}
se_linc <- se[LINC_g,]
heatmap_linc_data <- 
  se_linc 

heatmap_linc_data_m <-
  scale(assay(heatmap_linc_data, 2) %>% t() %>% log1p(), 
        center = TRUE, scale = TRUE) %>% t()
sample_site_color <- 
  colData(heatmap_linc_data) %>% as_tibble() %>% pull(biopsy_area) %>% unique()
color <- RColorBrewer::brewer.pal(length(sample_site_color), "Pastel2")
names(color) <- sample_site_color
heatmap_linc_annotation <- 
  HeatmapAnnotation(
    `Tissue type` = colData(heatmap_linc_data) %>% as_tibble() %>% pull(skin_type), 
    `Anatomic region` = colData(heatmap_linc_data) %>% as_tibble() %>% pull(biopsy_area),
    col = list(`Tissue type` = pca_color,
               `Anatomic region` = color),
    annotation_legend_param = list(`Tissue type` = list(nrow = 1, 
                                                        title_gp = gpar(fontsize = 16),
                                                        labels_gp = gpar(fontsize = 16)),
                                   `Anatomic region` = list(nrow = 1,
                                                          title_gp = gpar(fontsize = 16),
                                                          labels_gp = gpar(fontsize = 16))
                                   ))

library(dendsort)
col_dend <- heatmap_linc_data_m %>% t() %>% dist() %>% hclust
col_dend_sort <- dendsort(col_dend)

heatmap_linc <- ComplexHeatmap::Heatmap(heatmap_linc_data_m,
                        name = "gene expression",
                        column_dend_reorder = TRUE,
                        row_dend_reorder = TRUE,
                        show_column_names = FALSE,
                        show_row_names = T,
                        show_row_dend = FALSE,
                        # cluster_columns = col_dend_sort,
                        row_names_gp = gpar(fontsize = min(8, 800 * 6 /dim(heatmap_linc_data)[1])),
                        top_annotation = heatmap_linc_annotation,
                        heatmap_legend_param = list(
                          title_gp = gpar(fontsize = 16),
                          labels_gp = gpar(fontsize = 14)
                          ))

draw(heatmap_linc, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)
```

```{r}
png("data/supplementary/figure_s5.png", 
    width = 480*6, height = 480*11, res=300)
draw(heatmap_linc, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)
dev.off()
```

```{r heatmap-mir}
se_mir <- se[MIR_g,]
heatmap_mir_data <- 
  se_mir
heatmap_mir_data_m <-
  scale(assay(heatmap_mir_data, 2) %>% t() %>% log1p(), 
        center = TRUE, scale = TRUE) %>% t()
sample_site_color <- 
  colData(heatmap_mir_data) %>% as_tibble() %>% pull(biopsy_area) %>% unique()
color <- RColorBrewer::brewer.pal(length(sample_site_color), "Pastel2")
names(color) <- sample_site_color
heatmap_mir_annotation <- 
  HeatmapAnnotation(
    `Tissue type` = colData(heatmap_mir_data) %>% as_tibble() %>% pull(skin_type), 
    `Anatomic region` = colData(heatmap_mir_data) %>% as_tibble() %>% pull(biopsy_area),
    col = list(`Tissue type` = pca_color,
               `Anatomic region` = color),
    # simple_anno_size = unit(1, "cm"),
    annotation_legend_param = list(`Tissue type` = list(nrow = 1, 
                                                        # grid_height = unit(2, "cm"), 
                                                        # grid_width = unit(1, "cm"), 
                                                        title_gp = gpar(fontsize = 16),
                                                        labels_gp = gpar(fontsize = 16)),
                                   `Anatomic region` = list(nrow = 1,
                                                          # grid_height = unit(2, "cm"), 
                                                          # grid_width = unit(1, "cm"), 
                                                          title_gp = gpar(fontsize = 16),
                                                          labels_gp = gpar(fontsize = 16))
                                   ))


heatmap_mir <- ComplexHeatmap::Heatmap(heatmap_mir_data_m,
                        name = "gene expression",
                        # column_dend_reorder = TRUE,
                        # row_dend_reorder = TRUE,
                        # clustering_method_columns = "complete",
                        cluster_columns = heatmap_mir_data_m %>% t() %>% dist() %>% hclust() %>% dendsort(),
                        cluster_rows = heatmap_mir_data_m %>% dist() %>% hclust() %>% dendsort(),
                        show_column_names = FALSE,
                        show_row_names = T,
                        show_row_dend = FALSE,
                        row_names_gp = gpar(fontsize = 10),
                        top_annotation = heatmap_mir_annotation,
                        heatmap_legend_param = list(
                          title_gp = gpar(fontsize = 16),
                          labels_gp = gpar(fontsize = 14)
                          ))

draw(heatmap_mir, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)
```

```{r save heatmap}
png("data/supplementary/figure_s2_c.png", 
    width = 480*6.5, 
    height = 480*4, 
    res=300)
draw(heatmap_mir, heatmap_legend_side = "right",
     annotation_legend_side = "top", merge_legend = FALSE)
dev.off()
```

```{r MIR31 and CLDN1}
MIR31_CLDN1 <- 
  se[c("MIR31HG", "CLDN1"),] %>% as_tibble() %>% 
  filter(skin_type == "LS") %>% 
  select(feature, sample, counts_scaled, skin_type) %>% pivot_wider(names_from = feature, values_from = counts_scaled) %>% 
  ggscatter(x = "CLDN1", 
            y = "MIR31HG",
            ylab = "MIR31HG",
            xscale = "log10",
            add = "reg.line",
            ) +
  stat_cor()
ggsave("data/supplementary/figure_s2_d.png", 
       MIR31_CLDN1, width = 7, height = 5, dpi = 300)
```
